{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0. Tools and Python (3.7.4) packages   \n",
    "**bedtools v2.31.0**  \n",
    "**pandas v1.3.5**   \n",
    "**NumPy v1.21.5**  \n",
    "**Bio v1.5.9** \n",
    "\n",
    "# Resources   \n",
    "1. [UCSC Genome Browser](http://genome.ucsc.edu/index.html)   \n",
    "2. [NCBI](https://www.ncbi.nlm.nih.gov/)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import Entrez\n",
    "Entrez.email = \"an.eudokimova@yandex.ru\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "****************"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Task1. Annotation of input BED-file   \n",
    "1.1 Download GTF files for the main gene transcript sets from [[1]]. The BED-file header states that reference=hg19:     \n",
    "```bash\n",
    "wget https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/genes/hg19.ncbiRefSeq.gtf.gz\n",
    "```   \n",
    "```bash \n",
    "gunzip hg19.ncbiRefSeq.gtf.gz\n",
    "```   \n",
    "   \n",
    "Based on the wording of the task, we are only interested in exons, so we extract only exons from GTF-file:   \n",
    "```bash \n",
    "grep -w 'exon' hg19.ncbiRefSeq.gtf > hg19.ncbiRefSeq.exon.gtf\n",
    "```\n",
    "   \n",
    "1.2 To identify gene regions in input BED-file we will use bedtools intersect, this tool allows one to screen for overlaps between two sets of genomic features:   \n",
    "``` bash \n",
    "bedtools intersect -a IAD143293_241_Designed.bed -b hg19.ncbiRefSeq.exon.gtf -loj -wa -wb > intersect.txt\n",
    "```   \n",
    "Options `-wa` and `-wb` allows to write the original entries of both A and B input files in output   \n",
    "Option `-loj` report a NULL feature for B, if no overlaps are found   \n",
    "\n",
    "Creat directory for final results:   \n",
    "```bash \n",
    "mkdir RESULTS\n",
    "```\n",
    "\n",
    "[1]: https://hgdownload.soe.ucsc.edu/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1.3 Analisys of output file   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18722630</td>\n",
       "      <td>18722936</td>\n",
       "      <td>AMPL7162509462</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7162509462</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113456395</td>\n",
       "      <td>113456721</td>\n",
       "      <td>AMPL7166063003</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166063003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113454469</td>\n",
       "      <td>113456787</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SLC16A1\"; transcript_id \"NM_001166496...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113456395</td>\n",
       "      <td>113456721</td>\n",
       "      <td>AMPL7166063003</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166063003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113454469</td>\n",
       "      <td>113456787</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SLC16A1\"; transcript_id \"NM_003051.4\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113459897</td>\n",
       "      <td>113460203</td>\n",
       "      <td>AMPL7166063010</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063010</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113459800</td>\n",
       "      <td>113460666</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SLC16A1\"; transcript_id \"NM_001166496...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113459897</td>\n",
       "      <td>113460203</td>\n",
       "      <td>AMPL7166063010</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063010</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113459800</td>\n",
       "      <td>113460666</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SLC16A1\"; transcript_id \"NM_003051.4\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0          1          2               3  4   \\\n",
       "0  chr1   18722630   18722936  AMPL7162509462  .   \n",
       "1  chr1  113456395  113456721  AMPL7166063003  .   \n",
       "2  chr1  113456395  113456721  AMPL7166063003  .   \n",
       "3  chr1  113459897  113460203  AMPL7166063010  .   \n",
       "4  chr1  113459897  113460203  AMPL7166063010  .   \n",
       "\n",
       "                                       5     6                      7     8   \\\n",
       "0  Pool=1;SUBMITTED_REGION=AMPL7162509462     .                      .     .   \n",
       "1  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "2  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "3  Pool=1;SUBMITTED_REGION=AMPL7166063010  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "4  Pool=1;SUBMITTED_REGION=AMPL7166063010  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "\n",
       "          9          10 11 12 13  \\\n",
       "0         -1         -1  .  .  .   \n",
       "1  113454469  113456787  .  -  .   \n",
       "2  113454469  113456787  .  -  .   \n",
       "3  113459800  113460666  .  -  .   \n",
       "4  113459800  113460666  .  -  .   \n",
       "\n",
       "                                                  14  \n",
       "0                                                  .  \n",
       "1  gene_id \"SLC16A1\"; transcript_id \"NM_001166496...  \n",
       "2  gene_id \"SLC16A1\"; transcript_id \"NM_003051.4\"...  \n",
       "3  gene_id \"SLC16A1\"; transcript_id \"NM_001166496...  \n",
       "4  gene_id \"SLC16A1\"; transcript_id \"NM_003051.4\"...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect = pd.read_csv('intersect.txt', sep='\\t', header=None)\n",
    "intersect.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see, that not all amplicons are identified (-1 in columns 9 and 10).    \n",
    "Extraction gene name and exon number from column 14:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18722630</td>\n",
       "      <td>18722936</td>\n",
       "      <td>AMPL7162509462</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7162509462</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113456395</td>\n",
       "      <td>113456721</td>\n",
       "      <td>AMPL7166063003</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166063003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113454469</td>\n",
       "      <td>113456787</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SLC16A1\"; transcript_id \"NM_001166496...</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113456395</td>\n",
       "      <td>113456721</td>\n",
       "      <td>AMPL7166063003</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166063003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113454469</td>\n",
       "      <td>113456787</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SLC16A1\"; transcript_id \"NM_003051.4\"...</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113459897</td>\n",
       "      <td>113460203</td>\n",
       "      <td>AMPL7166063010</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063010</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113459800</td>\n",
       "      <td>113460666</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SLC16A1\"; transcript_id \"NM_001166496...</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113459897</td>\n",
       "      <td>113460203</td>\n",
       "      <td>AMPL7166063010</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063010</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113459800</td>\n",
       "      <td>113460666</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SLC16A1\"; transcript_id \"NM_003051.4\"...</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113460270</td>\n",
       "      <td>113460593</td>\n",
       "      <td>AMPL7163702396</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7163702396</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113459800</td>\n",
       "      <td>113460666</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SLC16A1\"; transcript_id \"NM_001166496...</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     0          1          2               3  4   \\\n",
       "0  chr1   18722630   18722936  AMPL7162509462  .   \n",
       "1  chr1  113456395  113456721  AMPL7166063003  .   \n",
       "2  chr1  113456395  113456721  AMPL7166063003  .   \n",
       "3  chr1  113459897  113460203  AMPL7166063010  .   \n",
       "4  chr1  113459897  113460203  AMPL7166063010  .   \n",
       "5  chr1  113460270  113460593  AMPL7163702396  .   \n",
       "\n",
       "                                       5     6                      7     8   \\\n",
       "0  Pool=1;SUBMITTED_REGION=AMPL7162509462     .                      .     .   \n",
       "1  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "2  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "3  Pool=1;SUBMITTED_REGION=AMPL7166063010  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "4  Pool=1;SUBMITTED_REGION=AMPL7166063010  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "5  Pool=2;SUBMITTED_REGION=AMPL7163702396  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "\n",
       "          9          10 11 12 13  \\\n",
       "0         -1         -1  .  .  .   \n",
       "1  113454469  113456787  .  -  .   \n",
       "2  113454469  113456787  .  -  .   \n",
       "3  113459800  113460666  .  -  .   \n",
       "4  113459800  113460666  .  -  .   \n",
       "5  113459800  113460666  .  -  .   \n",
       "\n",
       "                                                  14       15 16  \n",
       "0                                                  .        .  .  \n",
       "1  gene_id \"SLC16A1\"; transcript_id \"NM_001166496...  SLC16A1  5  \n",
       "2  gene_id \"SLC16A1\"; transcript_id \"NM_003051.4\"...  SLC16A1  5  \n",
       "3  gene_id \"SLC16A1\"; transcript_id \"NM_001166496...  SLC16A1  4  \n",
       "4  gene_id \"SLC16A1\"; transcript_id \"NM_003051.4\"...  SLC16A1  4  \n",
       "5  gene_id \"SLC16A1\"; transcript_id \"NM_001166496...  SLC16A1  4  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = intersect[14].str.split(' ', expand=True).iloc[:, [9, 5]]. \\\n",
    "    reset_index().drop('index', axis=1).replace(np.nan, '.').replace([';', '\"'], '', regex=True)\n",
    "tmp.columns = [15, 16]\n",
    "out = pd.concat([intersect, tmp], axis=1)\n",
    "out.head(6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see rows with the same amplicons, gene names and exon coordinates and number, but with different transcript_id in column 14. As far as we are not interested in transcript_id, we can remove this duplicates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "out = out.drop_duplicates([3, 15, 16], keep='first').reset_index().drop('index', axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "...and finaly extract target columns, rename them and write output file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHR</th>\n",
       "      <th>START_POS</th>\n",
       "      <th>END_POS</th>\n",
       "      <th>AMPL_ID</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>AMPL_DESCRIPTION</th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>EXON_NUMBER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>18722630</td>\n",
       "      <td>18722936</td>\n",
       "      <td>AMPL7162509462</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7162509462</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113456395</td>\n",
       "      <td>113456721</td>\n",
       "      <td>AMPL7166063003</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166063003</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113459897</td>\n",
       "      <td>113460203</td>\n",
       "      <td>AMPL7166063010</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063010</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113460270</td>\n",
       "      <td>113460593</td>\n",
       "      <td>AMPL7163702396</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7163702396</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113471727</td>\n",
       "      <td>113472054</td>\n",
       "      <td>AMPL7166063018</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063018</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113498778</td>\n",
       "      <td>113499101</td>\n",
       "      <td>AMPL7156752826</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7156752826</td>\n",
       "      <td>SLC16A1-AS1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #CHR  START_POS    END_POS         AMPL_ID STRAND  \\\n",
       "0  chr1   18722630   18722936  AMPL7162509462      .   \n",
       "1  chr1  113456395  113456721  AMPL7166063003      .   \n",
       "2  chr1  113459897  113460203  AMPL7166063010      .   \n",
       "3  chr1  113460270  113460593  AMPL7163702396      .   \n",
       "4  chr1  113471727  113472054  AMPL7166063018      .   \n",
       "5  chr1  113498778  113499101  AMPL7156752826      .   \n",
       "\n",
       "                         AMPL_DESCRIPTION    GENE_NAME EXON_NUMBER  \n",
       "0  Pool=1;SUBMITTED_REGION=AMPL7162509462            .           .  \n",
       "1  Pool=2;SUBMITTED_REGION=AMPL7166063003      SLC16A1           5  \n",
       "2  Pool=1;SUBMITTED_REGION=AMPL7166063010      SLC16A1           4  \n",
       "3  Pool=2;SUBMITTED_REGION=AMPL7163702396      SLC16A1           4  \n",
       "4  Pool=1;SUBMITTED_REGION=AMPL7166063018      SLC16A1           2  \n",
       "5  Pool=2;SUBMITTED_REGION=AMPL7156752826  SLC16A1-AS1           1  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out = out.iloc[:, [0,1, 2, 3, 4, 5, 15, 16]]\n",
    "out.columns = ['#CHR', 'START_POS', 'END_POS', 'AMPL_ID', 'STRAND', 'AMPL_DESCRIPTION', 'GENE_NAME', 'EXON_NUMBER']\n",
    "out.to_csv('RESULTS/task1_output.bed', sep='\\t', index=None, header=True)\n",
    "out.head(6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add header of input BED-file in output file:   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('IAD143293_241_Designed.bed', 'r') as input:\n",
    "    header = input.readline()  \n",
    "with open('RESULTS/task1_output.bed', 'r+') as output:\n",
    "    content = output.read()\n",
    "    output.seek(0, 0)  \n",
    "    output.write(header)  \n",
    "    output.write(content)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1.4 Get information about genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function for searching gene ID, description and summary\n",
    "def get_id(gene):\n",
    "    try:\n",
    "        handle = Entrez.esearch(db=\"gene\",term=f\"Homo[Orgn] AND {gene}[Gene]\")\n",
    "        result = Entrez.read(handle)\n",
    "        to_return = result['IdList'][0]\n",
    "    except:\n",
    "        to_return = np.nan\n",
    "    return to_return\n",
    "\n",
    "def get_summary(ID):\n",
    "    try:\n",
    "        handle = Entrez.epost(\"gene\", id=ID)\n",
    "        result = Entrez.read(handle)\n",
    "        webEnv = result[\"WebEnv\"]\n",
    "        queryKey = result[\"QueryKey\"]\n",
    "        data = Entrez.esummary(db=\"gene\", webenv=webEnv, query_key=queryKey)\n",
    "        result = Entrez.read(data)\n",
    "        to_return = [result['DocumentSummarySet']['DocumentSummary'][0]['Description'],\n",
    "            result['DocumentSummarySet']['DocumentSummary'][0]['Summary']]\n",
    "    except:\n",
    "        to_return = [np.nan, np.nan]\n",
    "    return to_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create DF with gene names\n",
    "genes = pd.DataFrame({'GENE_NAME':out['GENE_NAME'].unique()})\n",
    "genes = genes[genes['GENE_NAME'] != '.'].reset_index().drop('index', axis=1)\n",
    "\n",
    "# get ID, description and summary of each gene using Entrez\n",
    "genes['ID'] = genes['GENE_NAME'].apply(get_id)\n",
    "tmp = pd.DataFrame(list(map(get_summary, genes['ID'])))\n",
    "tmp.columns = ['DESCRIPTION', 'SUMMARY']\n",
    "genes = pd.concat([genes, tmp], axis=1)\n",
    "\n",
    "genes.to_csv('RESULTS/genes_summary.csv', sep='\\t', index=None)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have 31 genes (see table below). Some functions of protein encoded by these genes:   \n",
    "- transpoters (*SLC16A1; ABCC8*)   \n",
    "- enzymes (*G6PC2; HADH*)   \n",
    "- transcription regulatory factors or part of them (*KLF11 - cell growth inhibition and apoptosis; NEUROD1 - regulation expression of the insulin gene, ZFP57; RFX6; PAX4; GLIS3; NEUROG3 - involved in neurogenesis; HNF1A - regulation of expression of several liver-specific genes; HNF4A; PDX1 - activator of insulin; somatostatin; glucokinase etc; HNF1B - regulation of nephron and embryonic pancreas development; FOXP3*)   \n",
    "- kinases (*GCK; BLK - cell proliferation and differentiation; GLUD1 - regulating amino acid-induced insulin secretion; INSR - insulin receptor*)   \n",
    "- potassium channel (*KCNJ11*)    \n",
    "    \n",
    "Based on information about these genes, I can assume, that this targeted panel under consideration can be used to diagnose and assess susceptibility to various types of diabetes, as well as other metabolic disorders associated with the functioning of the pancreas and liver.    \n",
    "Some of these genes can be used for canser diagnostic (for example *BLK, HNF1B, PAX4*), but I don't think that it is a purpose of this pannel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE_NAME</th>\n",
       "      <th>ID</th>\n",
       "      <th>DESCRIPTION</th>\n",
       "      <th>SUMMARY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>6566</td>\n",
       "      <td>solute carrier family 16 member 1</td>\n",
       "      <td>The protein encoded by this gene is a proton-linked monocarboxylate transporter that catalyzes the movement of many monocarboxylates, such as lactate and pyruvate, across the plasma membrane. Mutations in this gene are associated with erythrocyte lactate transporter defect. Alternatively spliced transcript variants have been found for this gene.[provided by RefSeq, Oct 2009]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SLC16A1-AS1</td>\n",
       "      <td>100506392</td>\n",
       "      <td>SLC16A1 antisense RNA 1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KLF11</td>\n",
       "      <td>8462</td>\n",
       "      <td>KLF transcription factor 11</td>\n",
       "      <td>The protein encoded by this gene is a zinc finger transcription factor that binds to SP1-like sequences in epsilon- and gamma-globin gene promoters. This binding inhibits cell growth and causes apoptosis. Defects in this gene are a cause of maturity-onset diabetes of the young type 7 (MODY7). Three transcript variants encoding two different isoforms have been found for this gene. [provided by RefSeq, Apr 2010]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G6PC2</td>\n",
       "      <td>57818</td>\n",
       "      <td>glucose-6-phosphatase catalytic subunit 2</td>\n",
       "      <td>This gene encodes an enzyme belonging to the glucose-6-phosphatase catalytic subunit family. These enzymes are part of a multicomponent integral membrane system that catalyzes the hydrolysis of glucose-6-phosphate, the terminal step in gluconeogenic and glycogenolytic pathways, allowing the release of glucose into the bloodstream. The family member encoded by this gene is found in pancreatic islets and does not exhibit phosphohydrolase activity, but it is a major target of cell-mediated autoimmunity in diabetes. Several alternatively spliced transcript variants of this gene have been described, but their biological validity has not been determined. [provided by RefSeq, Jul 2008]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NEUROD1</td>\n",
       "      <td>4760</td>\n",
       "      <td>neuronal differentiation 1</td>\n",
       "      <td>This gene encodes a member of the NeuroD family of basic helix-loop-helix (bHLH) transcription factors. The protein forms heterodimers with other bHLH proteins and activates transcription of genes that contain a specific DNA sequence known as the E-box. It regulates expression of the insulin gene, and mutations in this gene result in type II diabetes mellitus. [provided by RefSeq, Jul 2008]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MFF-DT</td>\n",
       "      <td>654841</td>\n",
       "      <td>MFF divergent transcript</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HADH</td>\n",
       "      <td>3030</td>\n",
       "      <td>hydroxyacyl-CoA dehydrogenase trifunctional multienzyme complex subunit alpha</td>\n",
       "      <td>This gene encodes the alpha subunit of the mitochondrial trifunctional protein, which catalyzes the last three steps of mitochondrial beta-oxidation of long chain fatty acids. The mitochondrial membrane-bound heterocomplex is composed of four alpha and four beta subunits, with the alpha subunit catalyzing the 3-hydroxyacyl-CoA dehydrogenase and enoyl-CoA hydratase activities. Mutations in this gene result in trifunctional protein deficiency or LCHAD deficiency. The genes of the alpha and beta subunits of the mitochondrial trifunctional protein are located adjacent to each other in the human genome in a head-to-head orientation. [provided by RefSeq, Jul 2008]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ZFP57</td>\n",
       "      <td>346171</td>\n",
       "      <td>ZFP57 zinc finger protein</td>\n",
       "      <td>The protein encoded by this gene is a zinc finger protein containing a KRAB domain. Studies in mouse suggest that this protein may function as a transcriptional repressor. Mutations in this gene have been associated with transient neonatal diabetes mellitus type 1 (TNDM1).[provided by RefSeq, Sep 2009]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RFX6</td>\n",
       "      <td>222546</td>\n",
       "      <td>regulatory factor X6</td>\n",
       "      <td>The nuclear protein encoded by this gene is a member of the regulatory factor X (RFX) family of transcription factors. Studies in mice suggest that this gene is specifically required for the differentiation of islet cells for the production of insulin, but not for the differentiation of pancreatic polypeptide-producing cells. It regulates the transcription factors involved in beta-cell maturation and function, thus, restricting the expression of the beta-cell differentiation and specification genes. Mutations in this gene are associated with Mitchell-Riley syndrome, which is characterized by neonatal diabetes with pancreatic hypoplasia, duodenal and jejunal atresia, and gall bladder agenesis.[provided by RefSeq, Sep 2010]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GCK</td>\n",
       "      <td>2645</td>\n",
       "      <td>glucokinase</td>\n",
       "      <td>This gene encodes a member of the hexokinase family of proteins. Hexokinases phosphorylate glucose to produce glucose-6-phosphate, the first step in most glucose metabolism pathways. In contrast to other forms of hexokinase, this enzyme is not inhibited by its product glucose-6-phosphate but remains active while glucose is abundant. The use of multiple promoters and alternative splicing of this gene result in distinct protein isoforms that exhibit tissue-specific expression in the pancreas and liver. In the pancreas, this enzyme plays a role in glucose-stimulated insulin secretion, while in the liver, this enzyme is important in glucose uptake and conversion to glycogen. Mutations in this gene that alter enzyme activity have been associated with multiple types of diabetes and hyperinsulinemic hypoglycemia. [provided by RefSeq, Aug 2017]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PAX4</td>\n",
       "      <td>5078</td>\n",
       "      <td>paired box 4</td>\n",
       "      <td>This gene is a member of the paired box (PAX) family of transcription factors. Members of this gene family typically contain a paired box domain, an octapeptide, and a paired-type homeodomain. These genes play critical roles during fetal development and cancer growth. The paired box 4 gene is involved in pancreatic islet development and mouse studies have demonstrated a role for this gene in differentiation of insulin-producing beta cells. [provided by RefSeq, Jul 2008]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BLK</td>\n",
       "      <td>640</td>\n",
       "      <td>BLK proto-oncogene, Src family tyrosine kinase</td>\n",
       "      <td>This gene encodes a nonreceptor tyrosine-kinase of the src family of proto-oncogenes that are typically involved in cell proliferation and differentiation. The protein has a role in B-cell receptor signaling and B-cell development. The protein also stimulates insulin synthesis and secretion in response to glucose and enhances the expression of several pancreatic beta-cell transcription factors. [provided by RefSeq, Aug 2010]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GLIS3</td>\n",
       "      <td>169792</td>\n",
       "      <td>GLIS family zinc finger 3</td>\n",
       "      <td>This gene is a member of the GLI-similar zinc finger protein family and encodes a nuclear protein with five C2H2-type zinc finger domains. This protein functions as both a repressor and activator of transcription and is specifically involved in the development of pancreatic beta cells, the thyroid, eye, liver and kidney. Mutations in this gene have been associated with neonatal diabetes and congenital hypothyroidism (NDH). Alternatively spliced variants that encode different protein isoforms have been described but the full-length nature of only two have been determined. [provided by RefSeq, Jul 2008]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GLIS3-AS1</td>\n",
       "      <td>84850</td>\n",
       "      <td>GLIS3 antisense RNA 1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CEL</td>\n",
       "      <td>1056</td>\n",
       "      <td>carboxyl ester lipase</td>\n",
       "      <td>The protein encoded by this gene is a glycoprotein secreted from the pancreas into the digestive tract and from the lactating mammary gland into human milk. The physiological role of this protein is in cholesterol and lipid-soluble vitamin ester hydrolysis and absorption. This encoded protein promotes large chylomicron production in the intestine. Also its presence in plasma suggests its interactions with cholesterol and oxidized lipoproteins to modulate the progression of atherosclerosis. In pancreatic tumoral cells, this encoded protein is thought to be sequestrated within the Golgi compartment and is probably not secreted. This gene contains a variable number of tandem repeat (VNTR) polymorphism in the coding region that may influence the function of the encoded protein. [provided by RefSeq, Jul 2008]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NEUROG3</td>\n",
       "      <td>50674</td>\n",
       "      <td>neurogenin 3</td>\n",
       "      <td>The protein encoded by this gene is a basic helix-loop-helix (bHLH) transcription factor involved in neurogenesis. The encoded protein likely acts as a heterodimer with another bHLH protein. Defects in this gene are a cause of congenital malabsorptive diarrhea 4 (DIAR4).[provided by RefSeq, May 2010]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GLUD1</td>\n",
       "      <td>2746</td>\n",
       "      <td>glutamate dehydrogenase 1</td>\n",
       "      <td>This gene encodes glutamate dehydrogenase, which is a mitochondrial matrix enzyme that catalyzes the oxidative deamination of glutamate to alpha-ketoglutarate and ammonia. This enzyme has an important role in regulating amino acid-induced insulin secretion. It is allosterically activated by ADP and inhibited by GTP and ATP. Activating mutations in this gene are a common cause of congenital hyperinsulinism. Alternative splicing of this gene results in multiple transcript variants. The related glutamate dehydrogenase 2 gene on the human X-chromosome originated from this gene via retrotransposition and encodes a soluble form of glutamate dehydrogenase. Related pseudogenes have been identified on chromosomes 10, 18 and X. [provided by RefSeq, Jan 2016]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>INS</td>\n",
       "      <td>3630</td>\n",
       "      <td>insulin</td>\n",
       "      <td>This gene encodes insulin, a peptide hormone that plays a vital role in the regulation of carbohydrate and lipid metabolism. After removal of the precursor signal peptide, proinsulin is post-translationally cleaved into three peptides: the B chain and A chain peptides, which are covalently linked via two disulfide bonds to form insulin, and C-peptide. Binding of insulin to the insulin receptor (INSR) stimulates glucose uptake. A multitude of mutant alleles with phenotypic effects have been identified, including insulin-dependent diabetes mellitus, permanent neonatal diabetes diabetes mellitus, maturity-onset diabetes of the young type 10 and hyperproinsulinemia. There is a read-through gene, INS-IGF2, which overlaps with this gene at the 5' region and with the IGF2 gene at the 3' region. [provided by RefSeq, May 2020]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>INS-IGF2</td>\n",
       "      <td>723961</td>\n",
       "      <td>INS-IGF2 readthrough</td>\n",
       "      <td>This locus includes two alternatively spliced read-through transcript variants which align to the INS gene in the 5' region and to the IGF2 gene in the 3' region. One transcript is predicted to encode a protein which shares the N-terminus with the INS protein but has a distinct and longer C-terminus, whereas the other transcript is a candidate for nonsense-mediated decay (NMD). The transcripts are imprinted and are paternally expressed in the limb and eye. [provided by RefSeq, Jul 2008]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>KCNJ11</td>\n",
       "      <td>3767</td>\n",
       "      <td>potassium inwardly rectifying channel subfamily J member 11</td>\n",
       "      <td>Potassium channels are present in most mammalian cells, where they participate in a wide range of physiologic responses. The protein encoded by this gene is an integral membrane protein and inward-rectifier type potassium channel. The encoded protein, which has a greater tendency to allow potassium to flow into a cell rather than out of a cell, is controlled by G-proteins and is found associated with the sulfonylurea receptor SUR. Mutations in this gene are a cause of familial persistent hyperinsulinemic hypoglycemia of infancy (PHHI), an autosomal recessive disorder characterized by unregulated insulin secretion. Defects in this gene may also contribute to autosomal dominant non-insulin-dependent diabetes mellitus type II (NIDDM), transient neonatal diabetes mellitus type 3 (TNDM3), and permanent neonatal diabetes mellitus (PNDM). Multiple alternatively spliced transcript variants that encode different protein isoforms have been described for this gene. [provided by RefSeq, Oct 2009]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ABCC8</td>\n",
       "      <td>6833</td>\n",
       "      <td>ATP binding cassette subfamily C member 8</td>\n",
       "      <td>The protein encoded by this gene is a member of the superfamily of ATP-binding cassette (ABC) transporters. ABC proteins transport various molecules across extra- and intra-cellular membranes. ABC genes are divided into seven distinct subfamilies (ABC1, MDR/TAP, MRP, ALD, OABP, GCN20, White). This protein is a member of the MRP subfamily which is involved in multi-drug resistance. This protein functions as a modulator of ATP-sensitive potassium channels and insulin release. Mutations in the ABCC8 gene and deficiencies in the encoded protein have been observed in patients with hyperinsulinemic hypoglycemia of infancy, an autosomal recessive disorder of unregulated and high insulin secretion. Mutations have also been associated with non-insulin-dependent diabetes mellitus type II, an autosomal dominant disease of defective insulin secretion. Alternatively spliced transcript variants have been found for this gene. [provided by RefSeq, Jul 2020]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HNF1A</td>\n",
       "      <td>6927</td>\n",
       "      <td>HNF1 homeobox A</td>\n",
       "      <td>The protein encoded by this gene is a transcription factor required for the expression of several liver-specific genes. The encoded protein functions as a homodimer and binds to the inverted palindrome 5'-GTTAATNATTAAC-3'. Defects in this gene are a cause of maturity onset diabetes of the young type 3 (MODY3) and also can result in the appearance of hepatic adenomas. Alternative splicing results in multiple transcript variants encoding different isoforms. [provided by RefSeq, Apr 2015]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>C12orf43</td>\n",
       "      <td>64897</td>\n",
       "      <td>chromosome 12 open reading frame 43</td>\n",
       "      <td>Predicted to be involved in Spemann organizer formation and negative regulation of Wnt signaling pathway. Predicted to be located in nuclear envelope. [provided by Alliance of Genome Resources, Apr 2022]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>PDX1</td>\n",
       "      <td>3651</td>\n",
       "      <td>pancreatic and duodenal homeobox 1</td>\n",
       "      <td>The protein encoded by this gene is a transcriptional activator of several genes, including insulin, somatostatin, glucokinase, islet amyloid polypeptide, and glucose transporter type 2. The encoded nuclear protein is involved in the early development of the pancreas and plays a major role in glucose-dependent regulation of insulin gene expression. Defects in this gene are a cause of pancreatic agenesis, which can lead to early-onset insulin-dependent diabetes mellitus (IDDM), as well as maturity onset diabetes of the young type 4 (MODY4). [provided by RefSeq, Aug 2017]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HNF1B</td>\n",
       "      <td>6928</td>\n",
       "      <td>HNF1 homeobox B</td>\n",
       "      <td>This gene encodes a member of the homeodomain-containing superfamily of transcription factors. The protein binds to DNA as either a homodimer, or a heterodimer with the related protein hepatocyte nuclear factor 1-alpha. The gene has been shown to function in nephron development, and regulates development of the embryonic pancreas. Mutations in this gene result in renal cysts and diabetes syndrome and noninsulin-dependent diabetes mellitus, and expression of this gene is altered in some types of cancer. Multiple transcript variants encoding different isoforms have been found for this gene.[provided by RefSeq, Sep 2009]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>INSR</td>\n",
       "      <td>3643</td>\n",
       "      <td>insulin receptor</td>\n",
       "      <td>This gene encodes a member of the receptor tyrosine kinase family of proteins. The encoded preproprotein is proteolytically processed to generate alpha and beta subunits that form a heterotetrameric receptor. Binding of insulin or other ligands to this receptor activates the insulin signaling pathway, which regulates glucose uptake and release, as well as the synthesis and storage of carbohydrates, lipids and protein. Mutations in this gene underlie the inherited severe insulin resistance syndromes including type A insulin resistance syndrome, Donohue syndrome and Rabson-Mendenhall syndrome. Alternative splicing results in multiple transcript variants. [provided by RefSeq, Oct 2015]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NKX2-2</td>\n",
       "      <td>4821</td>\n",
       "      <td>NK2 homeobox 2</td>\n",
       "      <td>The protein encoded by this gene contains a homeobox domain and may be involved in the morphogenesis of the central nervous system. This gene is found on chromosome 20 near NKX2-4, and these two genes appear to be duplicated on chromosome 14 in the form of TITF1 and NKX2-8. The encoded protein is likely to be a nuclear transcription factor. [provided by RefSeq, Jul 2008]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HNF4A</td>\n",
       "      <td>3172</td>\n",
       "      <td>hepatocyte nuclear factor 4 alpha</td>\n",
       "      <td>The protein encoded by this gene is a nuclear transcription factor which binds DNA as a homodimer. The encoded protein controls the expression of several genes, including hepatocyte nuclear factor 1 alpha, a transcription factor which regulates the expression of several hepatic genes. This gene may play a role in development of the liver, kidney, and intestines. Mutations in this gene have been associated with monogenic autosomal dominant non-insulin-dependent diabetes mellitus type I. Alternative splicing of this gene results in multiple transcript variants encoding several different isoforms. [provided by RefSeq, Apr 2012]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>DBNDD2</td>\n",
       "      <td>55861</td>\n",
       "      <td>dysbindin domain containing 2</td>\n",
       "      <td>Involved in negative regulation of protein kinase activity. Predicted to be located in cytoplasm. [provided by Alliance of Genome Resources, Apr 2022]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>FOXP3</td>\n",
       "      <td>50943</td>\n",
       "      <td>forkhead box P3</td>\n",
       "      <td>The protein encoded by this gene is a member of the forkhead/winged-helix family of transcriptional regulators. Defects in this gene are the cause of immunodeficiency polyendocrinopathy, enteropathy, X-linked syndrome (IPEX), also known as X-linked autoimmunity-immunodeficiency syndrome. Alternatively spliced transcript variants encoding different isoforms have been identified. [provided by RefSeq, Jul 2008]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>UTY</td>\n",
       "      <td>7404</td>\n",
       "      <td>ubiquitously transcribed tetratricopeptide repeat containing, Y-linked</td>\n",
       "      <td>This gene encodes a protein containing tetratricopeptide repeats which are thought to be involved in protein-protein interactions. The encoded protein is also a minor histocompatibility antigen which may induce graft rejection of male stem cell grafts. A large number of alternatively spliced transcripts have been observed for this gene, but the full length nature of some of these variants has not been determined. [provided by RefSeq, Apr 2012]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      GENE_NAME         ID  \\\n",
       "0       SLC16A1       6566   \n",
       "1   SLC16A1-AS1  100506392   \n",
       "2         KLF11       8462   \n",
       "3         G6PC2      57818   \n",
       "4       NEUROD1       4760   \n",
       "5        MFF-DT     654841   \n",
       "6          HADH       3030   \n",
       "7         ZFP57     346171   \n",
       "8          RFX6     222546   \n",
       "9           GCK       2645   \n",
       "10         PAX4       5078   \n",
       "11          BLK        640   \n",
       "12        GLIS3     169792   \n",
       "13    GLIS3-AS1      84850   \n",
       "14          CEL       1056   \n",
       "15      NEUROG3      50674   \n",
       "16        GLUD1       2746   \n",
       "17          INS       3630   \n",
       "18     INS-IGF2     723961   \n",
       "19       KCNJ11       3767   \n",
       "20        ABCC8       6833   \n",
       "21        HNF1A       6927   \n",
       "22     C12orf43      64897   \n",
       "23         PDX1       3651   \n",
       "24        HNF1B       6928   \n",
       "25         INSR       3643   \n",
       "26       NKX2-2       4821   \n",
       "27        HNF4A       3172   \n",
       "28       DBNDD2      55861   \n",
       "29        FOXP3      50943   \n",
       "30          UTY       7404   \n",
       "\n",
       "                                                                      DESCRIPTION  \\\n",
       "0                                               solute carrier family 16 member 1   \n",
       "1                                                         SLC16A1 antisense RNA 1   \n",
       "2                                                     KLF transcription factor 11   \n",
       "3                                       glucose-6-phosphatase catalytic subunit 2   \n",
       "4                                                      neuronal differentiation 1   \n",
       "5                                                        MFF divergent transcript   \n",
       "6   hydroxyacyl-CoA dehydrogenase trifunctional multienzyme complex subunit alpha   \n",
       "7                                                       ZFP57 zinc finger protein   \n",
       "8                                                            regulatory factor X6   \n",
       "9                                                                     glucokinase   \n",
       "10                                                                   paired box 4   \n",
       "11                                 BLK proto-oncogene, Src family tyrosine kinase   \n",
       "12                                                      GLIS family zinc finger 3   \n",
       "13                                                          GLIS3 antisense RNA 1   \n",
       "14                                                          carboxyl ester lipase   \n",
       "15                                                                   neurogenin 3   \n",
       "16                                                      glutamate dehydrogenase 1   \n",
       "17                                                                        insulin   \n",
       "18                                                           INS-IGF2 readthrough   \n",
       "19                    potassium inwardly rectifying channel subfamily J member 11   \n",
       "20                                      ATP binding cassette subfamily C member 8   \n",
       "21                                                                HNF1 homeobox A   \n",
       "22                                            chromosome 12 open reading frame 43   \n",
       "23                                             pancreatic and duodenal homeobox 1   \n",
       "24                                                                HNF1 homeobox B   \n",
       "25                                                               insulin receptor   \n",
       "26                                                                 NK2 homeobox 2   \n",
       "27                                              hepatocyte nuclear factor 4 alpha   \n",
       "28                                                  dysbindin domain containing 2   \n",
       "29                                                                forkhead box P3   \n",
       "30         ubiquitously transcribed tetratricopeptide repeat containing, Y-linked   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    SUMMARY  \n",
       "0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 The protein encoded by this gene is a proton-linked monocarboxylate transporter that catalyzes the movement of many monocarboxylates, such as lactate and pyruvate, across the plasma membrane. Mutations in this gene are associated with erythrocyte lactate transporter defect. Alternatively spliced transcript variants have been found for this gene.[provided by RefSeq, Oct 2009]  \n",
       "1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             The protein encoded by this gene is a zinc finger transcription factor that binds to SP1-like sequences in epsilon- and gamma-globin gene promoters. This binding inhibits cell growth and causes apoptosis. Defects in this gene are a cause of maturity-onset diabetes of the young type 7 (MODY7). Three transcript variants encoding two different isoforms have been found for this gene. [provided by RefSeq, Apr 2010]  \n",
       "3                                                                                                                                                                                                                                                                                                                           This gene encodes an enzyme belonging to the glucose-6-phosphatase catalytic subunit family. These enzymes are part of a multicomponent integral membrane system that catalyzes the hydrolysis of glucose-6-phosphate, the terminal step in gluconeogenic and glycogenolytic pathways, allowing the release of glucose into the bloodstream. The family member encoded by this gene is found in pancreatic islets and does not exhibit phosphohydrolase activity, but it is a major target of cell-mediated autoimmunity in diabetes. Several alternatively spliced transcript variants of this gene have been described, but their biological validity has not been determined. [provided by RefSeq, Jul 2008]  \n",
       "4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 This gene encodes a member of the NeuroD family of basic helix-loop-helix (bHLH) transcription factors. The protein forms heterodimers with other bHLH proteins and activates transcription of genes that contain a specific DNA sequence known as the E-box. It regulates expression of the insulin gene, and mutations in this gene result in type II diabetes mellitus. [provided by RefSeq, Jul 2008]  \n",
       "5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            \n",
       "6                                                                                                                                                                                                                                                                                                                                                This gene encodes the alpha subunit of the mitochondrial trifunctional protein, which catalyzes the last three steps of mitochondrial beta-oxidation of long chain fatty acids. The mitochondrial membrane-bound heterocomplex is composed of four alpha and four beta subunits, with the alpha subunit catalyzing the 3-hydroxyacyl-CoA dehydrogenase and enoyl-CoA hydratase activities. Mutations in this gene result in trifunctional protein deficiency or LCHAD deficiency. The genes of the alpha and beta subunits of the mitochondrial trifunctional protein are located adjacent to each other in the human genome in a head-to-head orientation. [provided by RefSeq, Jul 2008]  \n",
       "7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           The protein encoded by this gene is a zinc finger protein containing a KRAB domain. Studies in mouse suggest that this protein may function as a transcriptional repressor. Mutations in this gene have been associated with transient neonatal diabetes mellitus type 1 (TNDM1).[provided by RefSeq, Sep 2009]  \n",
       "8                                                                                                                                                                                                                                                                               The nuclear protein encoded by this gene is a member of the regulatory factor X (RFX) family of transcription factors. Studies in mice suggest that this gene is specifically required for the differentiation of islet cells for the production of insulin, but not for the differentiation of pancreatic polypeptide-producing cells. It regulates the transcription factors involved in beta-cell maturation and function, thus, restricting the expression of the beta-cell differentiation and specification genes. Mutations in this gene are associated with Mitchell-Riley syndrome, which is characterized by neonatal diabetes with pancreatic hypoplasia, duodenal and jejunal atresia, and gall bladder agenesis.[provided by RefSeq, Sep 2010]  \n",
       "9                                                                                                                                                          This gene encodes a member of the hexokinase family of proteins. Hexokinases phosphorylate glucose to produce glucose-6-phosphate, the first step in most glucose metabolism pathways. In contrast to other forms of hexokinase, this enzyme is not inhibited by its product glucose-6-phosphate but remains active while glucose is abundant. The use of multiple promoters and alternative splicing of this gene result in distinct protein isoforms that exhibit tissue-specific expression in the pancreas and liver. In the pancreas, this enzyme plays a role in glucose-stimulated insulin secretion, while in the liver, this enzyme is important in glucose uptake and conversion to glycogen. Mutations in this gene that alter enzyme activity have been associated with multiple types of diabetes and hyperinsulinemic hypoglycemia. [provided by RefSeq, Aug 2017]  \n",
       "10                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               This gene is a member of the paired box (PAX) family of transcription factors. Members of this gene family typically contain a paired box domain, an octapeptide, and a paired-type homeodomain. These genes play critical roles during fetal development and cancer growth. The paired box 4 gene is involved in pancreatic islet development and mouse studies have demonstrated a role for this gene in differentiation of insulin-producing beta cells. [provided by RefSeq, Jul 2008]  \n",
       "11                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             This gene encodes a nonreceptor tyrosine-kinase of the src family of proto-oncogenes that are typically involved in cell proliferation and differentiation. The protein has a role in B-cell receptor signaling and B-cell development. The protein also stimulates insulin synthesis and secretion in response to glucose and enhances the expression of several pancreatic beta-cell transcription factors. [provided by RefSeq, Aug 2010]  \n",
       "12                                                                                                                                                                                                                                                                                                                                                                                                         This gene is a member of the GLI-similar zinc finger protein family and encodes a nuclear protein with five C2H2-type zinc finger domains. This protein functions as both a repressor and activator of transcription and is specifically involved in the development of pancreatic beta cells, the thyroid, eye, liver and kidney. Mutations in this gene have been associated with neonatal diabetes and congenital hypothyroidism (NDH). Alternatively spliced variants that encode different protein isoforms have been described but the full-length nature of only two have been determined. [provided by RefSeq, Jul 2008]  \n",
       "13                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           \n",
       "14                                                                                                                                                                                          The protein encoded by this gene is a glycoprotein secreted from the pancreas into the digestive tract and from the lactating mammary gland into human milk. The physiological role of this protein is in cholesterol and lipid-soluble vitamin ester hydrolysis and absorption. This encoded protein promotes large chylomicron production in the intestine. Also its presence in plasma suggests its interactions with cholesterol and oxidized lipoproteins to modulate the progression of atherosclerosis. In pancreatic tumoral cells, this encoded protein is thought to be sequestrated within the Golgi compartment and is probably not secreted. This gene contains a variable number of tandem repeat (VNTR) polymorphism in the coding region that may influence the function of the encoded protein. [provided by RefSeq, Jul 2008]  \n",
       "15                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            The protein encoded by this gene is a basic helix-loop-helix (bHLH) transcription factor involved in neurogenesis. The encoded protein likely acts as a heterodimer with another bHLH protein. Defects in this gene are a cause of congenital malabsorptive diarrhea 4 (DIAR4).[provided by RefSeq, May 2010]  \n",
       "16                                                                                                                                                                                                                                                   This gene encodes glutamate dehydrogenase, which is a mitochondrial matrix enzyme that catalyzes the oxidative deamination of glutamate to alpha-ketoglutarate and ammonia. This enzyme has an important role in regulating amino acid-induced insulin secretion. It is allosterically activated by ADP and inhibited by GTP and ATP. Activating mutations in this gene are a common cause of congenital hyperinsulinism. Alternative splicing of this gene results in multiple transcript variants. The related glutamate dehydrogenase 2 gene on the human X-chromosome originated from this gene via retrotransposition and encodes a soluble form of glutamate dehydrogenase. Related pseudogenes have been identified on chromosomes 10, 18 and X. [provided by RefSeq, Jan 2016]  \n",
       "17                                                                                                                                                                            This gene encodes insulin, a peptide hormone that plays a vital role in the regulation of carbohydrate and lipid metabolism. After removal of the precursor signal peptide, proinsulin is post-translationally cleaved into three peptides: the B chain and A chain peptides, which are covalently linked via two disulfide bonds to form insulin, and C-peptide. Binding of insulin to the insulin receptor (INSR) stimulates glucose uptake. A multitude of mutant alleles with phenotypic effects have been identified, including insulin-dependent diabetes mellitus, permanent neonatal diabetes diabetes mellitus, maturity-onset diabetes of the young type 10 and hyperproinsulinemia. There is a read-through gene, INS-IGF2, which overlaps with this gene at the 5' region and with the IGF2 gene at the 3' region. [provided by RefSeq, May 2020]  \n",
       "18                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              This locus includes two alternatively spliced read-through transcript variants which align to the INS gene in the 5' region and to the IGF2 gene in the 3' region. One transcript is predicted to encode a protein which shares the N-terminus with the INS protein but has a distinct and longer C-terminus, whereas the other transcript is a candidate for nonsense-mediated decay (NMD). The transcripts are imprinted and are paternally expressed in the limb and eye. [provided by RefSeq, Jul 2008]  \n",
       "19  Potassium channels are present in most mammalian cells, where they participate in a wide range of physiologic responses. The protein encoded by this gene is an integral membrane protein and inward-rectifier type potassium channel. The encoded protein, which has a greater tendency to allow potassium to flow into a cell rather than out of a cell, is controlled by G-proteins and is found associated with the sulfonylurea receptor SUR. Mutations in this gene are a cause of familial persistent hyperinsulinemic hypoglycemia of infancy (PHHI), an autosomal recessive disorder characterized by unregulated insulin secretion. Defects in this gene may also contribute to autosomal dominant non-insulin-dependent diabetes mellitus type II (NIDDM), transient neonatal diabetes mellitus type 3 (TNDM3), and permanent neonatal diabetes mellitus (PNDM). Multiple alternatively spliced transcript variants that encode different protein isoforms have been described for this gene. [provided by RefSeq, Oct 2009]  \n",
       "20                                              The protein encoded by this gene is a member of the superfamily of ATP-binding cassette (ABC) transporters. ABC proteins transport various molecules across extra- and intra-cellular membranes. ABC genes are divided into seven distinct subfamilies (ABC1, MDR/TAP, MRP, ALD, OABP, GCN20, White). This protein is a member of the MRP subfamily which is involved in multi-drug resistance. This protein functions as a modulator of ATP-sensitive potassium channels and insulin release. Mutations in the ABCC8 gene and deficiencies in the encoded protein have been observed in patients with hyperinsulinemic hypoglycemia of infancy, an autosomal recessive disorder of unregulated and high insulin secretion. Mutations have also been associated with non-insulin-dependent diabetes mellitus type II, an autosomal dominant disease of defective insulin secretion. Alternatively spliced transcript variants have been found for this gene. [provided by RefSeq, Jul 2020]  \n",
       "21                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               The protein encoded by this gene is a transcription factor required for the expression of several liver-specific genes. The encoded protein functions as a homodimer and binds to the inverted palindrome 5'-GTTAATNATTAAC-3'. Defects in this gene are a cause of maturity onset diabetes of the young type 3 (MODY3) and also can result in the appearance of hepatic adenomas. Alternative splicing results in multiple transcript variants encoding different isoforms. [provided by RefSeq, Apr 2015]  \n",
       "22                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Predicted to be involved in Spemann organizer formation and negative regulation of Wnt signaling pathway. Predicted to be located in nuclear envelope. [provided by Alliance of Genome Resources, Apr 2022]  \n",
       "23                                                                                                                                                                                                                                                                                                                                                                                                                                         The protein encoded by this gene is a transcriptional activator of several genes, including insulin, somatostatin, glucokinase, islet amyloid polypeptide, and glucose transporter type 2. The encoded nuclear protein is involved in the early development of the pancreas and plays a major role in glucose-dependent regulation of insulin gene expression. Defects in this gene are a cause of pancreatic agenesis, which can lead to early-onset insulin-dependent diabetes mellitus (IDDM), as well as maturity onset diabetes of the young type 4 (MODY4). [provided by RefSeq, Aug 2017]  \n",
       "24                                                                                                                                                                                                                                                                                                                                                                                        This gene encodes a member of the homeodomain-containing superfamily of transcription factors. The protein binds to DNA as either a homodimer, or a heterodimer with the related protein hepatocyte nuclear factor 1-alpha. The gene has been shown to function in nephron development, and regulates development of the embryonic pancreas. Mutations in this gene result in renal cysts and diabetes syndrome and noninsulin-dependent diabetes mellitus, and expression of this gene is altered in some types of cancer. Multiple transcript variants encoding different isoforms have been found for this gene.[provided by RefSeq, Sep 2009]  \n",
       "25                                                                                                                                                                                                                                                                                                                      This gene encodes a member of the receptor tyrosine kinase family of proteins. The encoded preproprotein is proteolytically processed to generate alpha and beta subunits that form a heterotetrameric receptor. Binding of insulin or other ligands to this receptor activates the insulin signaling pathway, which regulates glucose uptake and release, as well as the synthesis and storage of carbohydrates, lipids and protein. Mutations in this gene underlie the inherited severe insulin resistance syndromes including type A insulin resistance syndrome, Donohue syndrome and Rabson-Mendenhall syndrome. Alternative splicing results in multiple transcript variants. [provided by RefSeq, Oct 2015]  \n",
       "26                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    The protein encoded by this gene contains a homeobox domain and may be involved in the morphogenesis of the central nervous system. This gene is found on chromosome 20 near NKX2-4, and these two genes appear to be duplicated on chromosome 14 in the form of TITF1 and NKX2-8. The encoded protein is likely to be a nuclear transcription factor. [provided by RefSeq, Jul 2008]  \n",
       "27                                                                                                                                                                                                                                                                                                                                                                                 The protein encoded by this gene is a nuclear transcription factor which binds DNA as a homodimer. The encoded protein controls the expression of several genes, including hepatocyte nuclear factor 1 alpha, a transcription factor which regulates the expression of several hepatic genes. This gene may play a role in development of the liver, kidney, and intestines. Mutations in this gene have been associated with monogenic autosomal dominant non-insulin-dependent diabetes mellitus type I. Alternative splicing of this gene results in multiple transcript variants encoding several different isoforms. [provided by RefSeq, Apr 2012]  \n",
       "28                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Involved in negative regulation of protein kinase activity. Predicted to be located in cytoplasm. [provided by Alliance of Genome Resources, Apr 2022]  \n",
       "29                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              The protein encoded by this gene is a member of the forkhead/winged-helix family of transcriptional regulators. Defects in this gene are the cause of immunodeficiency polyendocrinopathy, enteropathy, X-linked syndrome (IPEX), also known as X-linked autoimmunity-immunodeficiency syndrome. Alternatively spliced transcript variants encoding different isoforms have been identified. [provided by RefSeq, Jul 2008]  \n",
       "30                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          This gene encodes a protein containing tetratricopeptide repeats which are thought to be involved in protein-protein interactions. The encoded protein is also a minor histocompatibility antigen which may induce graft rejection of male stem cell grafts. A large number of alternatively spliced transcripts have been observed for this gene, but the full length nature of some of these variants has not been determined. [provided by RefSeq, Apr 2012]  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*********************"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Task2. Searching of off-target region in the pannel    \n",
    "2.1. First of all, let's **repeat BED-file annotation** whithout \"exon\" filter:   \n",
    "```bash\n",
    "bedtools intersect -a IAD143293_241_Designed.bed -b hg19.ncbiRefSeq.gtf -loj -wa -wb > intersect_full.txt\n",
    "```   \n",
    "Extract gene names and drop non-overlapping amplicons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113456395</td>\n",
       "      <td>113456721</td>\n",
       "      <td>AMPL7166063003</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166063003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113454469</td>\n",
       "      <td>113456787</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113456395</td>\n",
       "      <td>113456721</td>\n",
       "      <td>AMPL7166063003</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166063003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>113454469</td>\n",
       "      <td>113456512</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113456395</td>\n",
       "      <td>113456721</td>\n",
       "      <td>AMPL7166063003</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166063003</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>CDS</td>\n",
       "      <td>113456516</td>\n",
       "      <td>113456787</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>SLC16A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2               3  4  \\\n",
       "0  chr1  113456395  113456721  AMPL7166063003  .   \n",
       "1  chr1  113456395  113456721  AMPL7166063003  .   \n",
       "2  chr1  113456395  113456721  AMPL7166063003  .   \n",
       "\n",
       "                                        5     6                      7     8  \\\n",
       "0  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "1  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17  3UTR   \n",
       "2  Pool=2;SUBMITTED_REGION=AMPL7166063003  chr1  ncbiRefSeq.2021-05-17   CDS   \n",
       "\n",
       "           9         10 11 12 13       14  \n",
       "0  113454469  113456787  .  -  .  SLC16A1  \n",
       "1  113454469  113456512  .  -  .  SLC16A1  \n",
       "2  113456516  113456787  .  -  2  SLC16A1  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersect_full = pd.read_csv('intersect_full.txt', sep='\\t', header=None)\n",
    "tmp = intersect_full[14].str.split(' ', expand=True).iloc[:, 9]. \\\n",
    "    reset_index().drop('index', axis=1).replace(np.nan, '.').replace([';', '\"'], '', regex=True)\n",
    "# drop column 14 for shorter view below\n",
    "df = intersect_full.drop([14], axis=1)\n",
    "tmp.columns = [14]\n",
    "df = pd.concat([df, tmp], axis=1)\n",
    "df = df.loc[df[14] != '.'].reset_index().drop('index', axis=1)\n",
    "\n",
    "df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df[14].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.2 We still have 31 genes. Let's **find out, which amplicons mapped on two or more genes**: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AMPL7156752826': ['SLC16A1-AS1', 'SLC16A1'],\n",
       " 'AMPL7154848931': ['GLIS3-AS1', 'GLIS3'],\n",
       " 'AMPL7154784953': ['INS', 'INS-IGF2'],\n",
       " 'AMPL7154784954': ['INS', 'INS-IGF2'],\n",
       " 'AMPL7154784955': ['INS', 'INS-IGF2'],\n",
       " 'AMPL7154897326': ['INS', 'INS-IGF2'],\n",
       " 'AMPL7154511317': ['C12orf43', 'HNF1A'],\n",
       " 'AMPL7154784910': ['C12orf43', 'HNF1A'],\n",
       " 'AMPL7154784911': ['C12orf43', 'HNF1A'],\n",
       " 'AMPL7154784912': ['C12orf43', 'HNF1A'],\n",
       " 'AMPL7154783021': ['C12orf43', 'HNF1A'],\n",
       " 'AMPL7159709172': ['C12orf43', 'HNF1A'],\n",
       " 'AMPL7159709205': ['C12orf43', 'HNF1A']}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_interesting_genes(df):\n",
    "    ampls = df[3].unique()\n",
    "    d = {}\n",
    "    for ampl in ampls:\n",
    "        set_genes = list(df.loc[df[3] == ampl][14].unique())\n",
    "        if len(set_genes) > 1:\n",
    "            d[ampl] = set_genes\n",
    "    return d\n",
    "find_interesting_genes(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For next steps we take only these genes and amplicons:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp_d = find_interesting_genes(df)\n",
    "tmp_df = pd.DataFrame()\n",
    "for ampl in tmp_d.keys():\n",
    "    tmp_df = pd.concat([tmp_df, df.loc[(df[3] == ampl) & df[14].isin(tmp_d[ampl])]], axis=0)\n",
    "\n",
    "df = tmp_df.reset_index().drop('index', axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2.3 We already find information about them, but we can **check again, if any of genes are pseudogenes**. For this purpose we use [List] of pseudogenes in hg19. As we expected, there are no these genes in our target pannel.      \n",
    "But we already can see some \"strange\" genes in the table above: chromosome 12 open reading frame 43 *C12orf43*, *GLIS3* antisense RNA 1, *SLC16A1* antisense RNA 1, *MFF* divergent transcript. They could be potential off-tartget reogions.   \n",
    "     \n",
    "2.4 **Search for homologous regions - 1st way**   \n",
    "I have three ideas how to search how to search homologous or identical region, but only one I could implement. This is \"manualy\" method, and it is only suitable for searching for 100% homology.     \n",
    "     \n",
    "There are four possible position amplicon on target region (see the picture below): inside (<span style=\"color:blue\">**blue**</span>), partly on left side (<span style=\"color:green\">**green**</span>), partly on right side (<span style=\"color:yellow\">**yellow**</span> ), target inside amplicon (<span style=\"color:purple\">**purple**</span>).   \n",
    "    \n",
    "<img src=\"picture1.jpg\" width=\"100%\" height=\"50%\"  />      \n",
    "   \n",
    "We can verify this by calculating the lengths of the overlaps, they will be different and often different from the length of the amplicon. If one amplicon mapped on two or more different genome region, we can we can consider 100% homologous overlap sequencies in three cases:   \n",
    "- same left-side overlap  length  \n",
    "- same left-side overlap length   \n",
    "- same full overlap length  == amplicon lenght      \n",
    "   \n",
    "In <span style=\"color:purple\">**purple**</span> it's impossible to find homologous regions.   \n",
    "\n",
    "For further analysis we need to calculate amplicon and target region length, distance between amplicon and target region start positions, distance between amplicon and target region end positions:\n",
    "\n",
    "[List]: http://genome.ucsc.edu/cgi-bin/hgSearch?search=pseudogene+mRNA&hgsid=1712866984_aQKAtsLa46m4JJkhGbsoT2fVyJFe&db=hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>ampl_len</th>\n",
       "      <th>ref_len</th>\n",
       "      <th>ampl_start-ref_start</th>\n",
       "      <th>ref_end-ampl_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113498778</td>\n",
       "      <td>113499101</td>\n",
       "      <td>AMPL7156752826</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7156752826</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113499037</td>\n",
       "      <td>113499578</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1-AS1</td>\n",
       "      <td>323</td>\n",
       "      <td>541</td>\n",
       "      <td>-259</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113498778</td>\n",
       "      <td>113499101</td>\n",
       "      <td>AMPL7156752826</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7156752826</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113499037</td>\n",
       "      <td>113500344</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1-AS1</td>\n",
       "      <td>323</td>\n",
       "      <td>1307</td>\n",
       "      <td>-259</td>\n",
       "      <td>1243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113498778</td>\n",
       "      <td>113499101</td>\n",
       "      <td>AMPL7156752826</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7156752826</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113497898</td>\n",
       "      <td>113498818</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>323</td>\n",
       "      <td>920</td>\n",
       "      <td>880</td>\n",
       "      <td>-283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113498778</td>\n",
       "      <td>113499101</td>\n",
       "      <td>AMPL7156752826</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7156752826</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>5UTR</td>\n",
       "      <td>113497898</td>\n",
       "      <td>113498818</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>323</td>\n",
       "      <td>920</td>\n",
       "      <td>880</td>\n",
       "      <td>-283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113498778</td>\n",
       "      <td>113499101</td>\n",
       "      <td>AMPL7156752826</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7156752826</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113498657</td>\n",
       "      <td>113498818</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>323</td>\n",
       "      <td>161</td>\n",
       "      <td>121</td>\n",
       "      <td>-283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113498778</td>\n",
       "      <td>113499101</td>\n",
       "      <td>AMPL7156752826</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7156752826</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>5UTR</td>\n",
       "      <td>113498657</td>\n",
       "      <td>113498818</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>323</td>\n",
       "      <td>161</td>\n",
       "      <td>121</td>\n",
       "      <td>-283</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0          1          2               3  4  \\\n",
       "0  chr1  113498778  113499101  AMPL7156752826  .   \n",
       "1  chr1  113498778  113499101  AMPL7156752826  .   \n",
       "2  chr1  113498778  113499101  AMPL7156752826  .   \n",
       "3  chr1  113498778  113499101  AMPL7156752826  .   \n",
       "4  chr1  113498778  113499101  AMPL7156752826  .   \n",
       "5  chr1  113498778  113499101  AMPL7156752826  .   \n",
       "\n",
       "                                        5     6                      7     8  \\\n",
       "0  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "1  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "2  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "3  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  5UTR   \n",
       "4  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  exon   \n",
       "5  Pool=2;SUBMITTED_REGION=AMPL7156752826  chr1  ncbiRefSeq.2021-05-17  5UTR   \n",
       "\n",
       "           9         10 11 12 13           14  ampl_len  ref_len  \\\n",
       "0  113499037  113499578  .  +  .  SLC16A1-AS1       323      541   \n",
       "1  113499037  113500344  .  +  .  SLC16A1-AS1       323     1307   \n",
       "2  113497898  113498818  .  -  .      SLC16A1       323      920   \n",
       "3  113497898  113498818  .  -  .      SLC16A1       323      920   \n",
       "4  113498657  113498818  .  -  .      SLC16A1       323      161   \n",
       "5  113498657  113498818  .  -  .      SLC16A1       323      161   \n",
       "\n",
       "   ampl_start-ref_start  ref_end-ampl_end  \n",
       "0                  -259               477  \n",
       "1                  -259              1243  \n",
       "2                   880              -283  \n",
       "3                   880              -283  \n",
       "4                   121              -283  \n",
       "5                   121              -283  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['ampl_len'] = df[2] - df[1]\n",
    "df['ref_len'] = df[10] - df[9]\n",
    "df['ampl_start-ref_start'] = df[1] - df[9]\n",
    "df['ref_end-ampl_end'] = df[10] - df[2]\n",
    "\n",
    "df.head(6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If `start_ampl-start_ref` > 0 and `stop_ref-stop_ampl` > 0, it's <span style=\"color:blue\">**blue**</span> position (see picture above)   \n",
    "If `start_ampl-start_ref` < 0 and `stop_ref-stop_ampl` > 0, it's <span style=\"color:green\">**green**</span> position   \n",
    "If `start_ampl-start_ref` > 0 and `stop_ref-stop_ampl` < 0, it's <span style=\"color:yellow\">**yellow**</span> position    \n",
    "If `start_ampl-start_ref` < 0 and `stop_ref-stop_ampl` < 0, it's <span style=\"color:purple\">**purple**</span> position"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's calculate overlaps lengths, it will be negative in <span style=\"color:green\">**green**</span> positions, in case of to amplicon with same overlap length in different regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlap_length(df):\n",
    "    if (df['ampl_start-ref_start'] >= 0) and (df['ref_end-ampl_end'] >= 0):\n",
    "        return df['ampl_len']\n",
    "    elif (df['ampl_start-ref_start'] < 0) and (df['ref_end-ampl_end'] > 0):\n",
    "        return -(df['ampl_len'] + df['ampl_start-ref_start'])\n",
    "    elif (df['ampl_start-ref_start'] > 0) and (df['ref_end-ampl_end'] < 0):\n",
    "        return df['ampl_len'] + df['ref_end-ampl_end']\n",
    "    elif (df['ampl_start-ref_start'] <= 0) and (df['ref_end-ampl_end'] <= 0):\n",
    "        return np.nan\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['overlap_len'] = df.apply(lambda x: overlap_length(x), axis=1)\n",
    "df = df.dropna().reset_index().drop('index', axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can remove rows with the same amplicon ID, gene, part of gene and overlap length.    \n",
    "Also we can drop rows where `overlap_len` is unique, because as we notice before, 100% homologu should have same length:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop_duplicates([3, 8, 14, 'overlap_len'])\n",
    "\n",
    "def drop_unique_length(df):\n",
    "    tmp_list = []\n",
    "    for x in df['overlap_len'].unique():\n",
    "        n = list(df['overlap_len']).count(x)\n",
    "        if n > 1:\n",
    "            tmp_list.append(x)\n",
    "    return df.loc[(df['overlap_len'].isin(tmp_list))].reset_index().drop('index', axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>ampl_len</th>\n",
       "      <th>ref_len</th>\n",
       "      <th>ampl_start-ref_start</th>\n",
       "      <th>ref_end-ampl_end</th>\n",
       "      <th>overlap_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113498778</td>\n",
       "      <td>113499101</td>\n",
       "      <td>AMPL7156752826</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7156752826</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>113497898</td>\n",
       "      <td>113498818</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>323</td>\n",
       "      <td>920</td>\n",
       "      <td>880</td>\n",
       "      <td>-283</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>113498778</td>\n",
       "      <td>113499101</td>\n",
       "      <td>AMPL7156752826</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7156752826</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>5UTR</td>\n",
       "      <td>113497898</td>\n",
       "      <td>113498818</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SLC16A1</td>\n",
       "      <td>323</td>\n",
       "      <td>920</td>\n",
       "      <td>880</td>\n",
       "      <td>-283</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2181834</td>\n",
       "      <td>2182073</td>\n",
       "      <td>AMPL7154784953</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182439</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS</td>\n",
       "      <td>239</td>\n",
       "      <td>424</td>\n",
       "      <td>-181</td>\n",
       "      <td>366</td>\n",
       "      <td>-58.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2181834</td>\n",
       "      <td>2182073</td>\n",
       "      <td>AMPL7154784953</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182201</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>INS</td>\n",
       "      <td>239</td>\n",
       "      <td>186</td>\n",
       "      <td>-181</td>\n",
       "      <td>128</td>\n",
       "      <td>-58.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2181834</td>\n",
       "      <td>2182073</td>\n",
       "      <td>AMPL7154784953</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182218</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS-IGF2</td>\n",
       "      <td>239</td>\n",
       "      <td>203</td>\n",
       "      <td>-181</td>\n",
       "      <td>145</td>\n",
       "      <td>-58.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0          1          2               3  4  \\\n",
       "0   chr1  113498778  113499101  AMPL7156752826  .   \n",
       "1   chr1  113498778  113499101  AMPL7156752826  .   \n",
       "2  chr11    2181834    2182073  AMPL7154784953  .   \n",
       "3  chr11    2181834    2182073  AMPL7154784953  .   \n",
       "4  chr11    2181834    2182073  AMPL7154784953  .   \n",
       "\n",
       "                                                                      5  \\\n",
       "0                                Pool=2;SUBMITTED_REGION=AMPL7156752826   \n",
       "1                                Pool=2;SUBMITTED_REGION=AMPL7156752826   \n",
       "2  Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   \n",
       "3  Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   \n",
       "4  Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   \n",
       "\n",
       "       6                      7     8          9         10 11 12 13  \\\n",
       "0   chr1  ncbiRefSeq.2021-05-17  exon  113497898  113498818  .  -  .   \n",
       "1   chr1  ncbiRefSeq.2021-05-17  5UTR  113497898  113498818  .  -  .   \n",
       "2  chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182439  .  -  .   \n",
       "3  chr11  ncbiRefSeq.2021-05-17   CDS    2182015    2182201  .  -  0   \n",
       "4  chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   \n",
       "\n",
       "         14  ampl_len  ref_len  ampl_start-ref_start  ref_end-ampl_end  \\\n",
       "0   SLC16A1       323      920                   880              -283   \n",
       "1   SLC16A1       323      920                   880              -283   \n",
       "2       INS       239      424                  -181               366   \n",
       "3       INS       239      186                  -181               128   \n",
       "4  INS-IGF2       239      203                  -181               145   \n",
       "\n",
       "   overlap_len  \n",
       "0         40.0  \n",
       "1         40.0  \n",
       "2        -58.0  \n",
       "3        -58.0  \n",
       "4        -58.0  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = drop_unique_length(df)\n",
    "df.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice, that if we filter only exons, the set of \"interesting\" genes will be the same, so let's do it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SLC16A1' 'INS' 'INS-IGF2' 'C12orf43' 'HNF1A']\n",
      "['SLC16A1' 'INS' 'INS-IGF2' 'C12orf43' 'HNF1A']\n"
     ]
    }
   ],
   "source": [
    "print(df.loc[df[8] == 'exon'][14].unique())\n",
    "print(df[14].unique())\n",
    "\n",
    "df = df.loc[df[8] == 'exon'].reset_index().drop('index', axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And drop rows where `overlap_len` is unique again:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>ampl_len</th>\n",
       "      <th>ref_len</th>\n",
       "      <th>ampl_start-ref_start</th>\n",
       "      <th>ref_end-ampl_end</th>\n",
       "      <th>overlap_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2181834</td>\n",
       "      <td>2182073</td>\n",
       "      <td>AMPL7154784953</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182439</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS</td>\n",
       "      <td>239</td>\n",
       "      <td>424</td>\n",
       "      <td>-181</td>\n",
       "      <td>366</td>\n",
       "      <td>-58.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2181834</td>\n",
       "      <td>2182073</td>\n",
       "      <td>AMPL7154784953</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182218</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS-IGF2</td>\n",
       "      <td>239</td>\n",
       "      <td>203</td>\n",
       "      <td>-181</td>\n",
       "      <td>145</td>\n",
       "      <td>-58.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2182062</td>\n",
       "      <td>2182385</td>\n",
       "      <td>AMPL7154784954</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154784954,AMPL7154784953,AMPL7154784955</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182218</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS</td>\n",
       "      <td>323</td>\n",
       "      <td>203</td>\n",
       "      <td>47</td>\n",
       "      <td>-167</td>\n",
       "      <td>156.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2182062</td>\n",
       "      <td>2182385</td>\n",
       "      <td>AMPL7154784954</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154784954,AMPL7154784953,AMPL7154784955</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182218</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS-IGF2</td>\n",
       "      <td>323</td>\n",
       "      <td>203</td>\n",
       "      <td>47</td>\n",
       "      <td>-167</td>\n",
       "      <td>156.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2182130</td>\n",
       "      <td>2182458</td>\n",
       "      <td>AMPL7154784955</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326,AMPL7154784954</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182218</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS</td>\n",
       "      <td>328</td>\n",
       "      <td>203</td>\n",
       "      <td>115</td>\n",
       "      <td>-240</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2182130</td>\n",
       "      <td>2182458</td>\n",
       "      <td>AMPL7154784955</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326,AMPL7154784954</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182218</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS-IGF2</td>\n",
       "      <td>328</td>\n",
       "      <td>203</td>\n",
       "      <td>115</td>\n",
       "      <td>-240</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2182429</td>\n",
       "      <td>2182753</td>\n",
       "      <td>AMPL7154897326</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182015</td>\n",
       "      <td>2182439</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS</td>\n",
       "      <td>324</td>\n",
       "      <td>424</td>\n",
       "      <td>414</td>\n",
       "      <td>-314</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr11</td>\n",
       "      <td>2182429</td>\n",
       "      <td>2182753</td>\n",
       "      <td>AMPL7154897326</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326</td>\n",
       "      <td>chr11</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>2182398</td>\n",
       "      <td>2182439</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>INS-IGF2</td>\n",
       "      <td>324</td>\n",
       "      <td>41</td>\n",
       "      <td>31</td>\n",
       "      <td>-314</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121438967</td>\n",
       "      <td>121439264</td>\n",
       "      <td>AMPL7154784910</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154784910,AMPL7154511317,AMPL7154784911</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438289</td>\n",
       "      <td>121442292</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>C12orf43</td>\n",
       "      <td>297</td>\n",
       "      <td>4003</td>\n",
       "      <td>678</td>\n",
       "      <td>3028</td>\n",
       "      <td>297.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121438967</td>\n",
       "      <td>121439264</td>\n",
       "      <td>AMPL7154784910</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154784910,AMPL7154511317,AMPL7154784911</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438868</td>\n",
       "      <td>121440315</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>HNF1A</td>\n",
       "      <td>297</td>\n",
       "      <td>1447</td>\n",
       "      <td>99</td>\n",
       "      <td>1051</td>\n",
       "      <td>297.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121439250</td>\n",
       "      <td>121439535</td>\n",
       "      <td>AMPL7154784911</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154784910,AMPL7154784911</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438289</td>\n",
       "      <td>121442292</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>C12orf43</td>\n",
       "      <td>285</td>\n",
       "      <td>4003</td>\n",
       "      <td>961</td>\n",
       "      <td>2757</td>\n",
       "      <td>285.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121439250</td>\n",
       "      <td>121439535</td>\n",
       "      <td>AMPL7154784911</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154784910,AMPL7154784911</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438868</td>\n",
       "      <td>121440315</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>HNF1A</td>\n",
       "      <td>285</td>\n",
       "      <td>1447</td>\n",
       "      <td>382</td>\n",
       "      <td>780</td>\n",
       "      <td>285.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121439524</td>\n",
       "      <td>121439785</td>\n",
       "      <td>AMPL7154784912</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154784911,AMPL7154784912,AMPL7154783021</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438289</td>\n",
       "      <td>121442292</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>C12orf43</td>\n",
       "      <td>261</td>\n",
       "      <td>4003</td>\n",
       "      <td>1235</td>\n",
       "      <td>2507</td>\n",
       "      <td>261.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121439524</td>\n",
       "      <td>121439785</td>\n",
       "      <td>AMPL7154784912</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154784911,AMPL7154784912,AMPL7154783021</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438868</td>\n",
       "      <td>121440315</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>HNF1A</td>\n",
       "      <td>261</td>\n",
       "      <td>1447</td>\n",
       "      <td>656</td>\n",
       "      <td>530</td>\n",
       "      <td>261.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121439774</td>\n",
       "      <td>121440004</td>\n",
       "      <td>AMPL7154783021</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154783021,AMPL7159709172</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438289</td>\n",
       "      <td>121442292</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>C12orf43</td>\n",
       "      <td>230</td>\n",
       "      <td>4003</td>\n",
       "      <td>1485</td>\n",
       "      <td>2288</td>\n",
       "      <td>230.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121439774</td>\n",
       "      <td>121440004</td>\n",
       "      <td>AMPL7154783021</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154783021,AMPL7159709172</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438868</td>\n",
       "      <td>121440315</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>HNF1A</td>\n",
       "      <td>230</td>\n",
       "      <td>1447</td>\n",
       "      <td>906</td>\n",
       "      <td>311</td>\n",
       "      <td>230.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121439987</td>\n",
       "      <td>121440308</td>\n",
       "      <td>AMPL7159709172</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154783021,AMPL7159709205,AMPL7159709172</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438289</td>\n",
       "      <td>121442292</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>C12orf43</td>\n",
       "      <td>321</td>\n",
       "      <td>4003</td>\n",
       "      <td>1698</td>\n",
       "      <td>1984</td>\n",
       "      <td>321.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr12</td>\n",
       "      <td>121439987</td>\n",
       "      <td>121440308</td>\n",
       "      <td>AMPL7159709172</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154783021,AMPL7159709205,AMPL7159709172</td>\n",
       "      <td>chr12</td>\n",
       "      <td>ncbiRefSeq.2021-05-17</td>\n",
       "      <td>exon</td>\n",
       "      <td>121438868</td>\n",
       "      <td>121440315</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>HNF1A</td>\n",
       "      <td>321</td>\n",
       "      <td>1447</td>\n",
       "      <td>1119</td>\n",
       "      <td>7</td>\n",
       "      <td>321.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0          1          2               3  4  \\\n",
       "0   chr11    2181834    2182073  AMPL7154784953  .   \n",
       "1   chr11    2181834    2182073  AMPL7154784953  .   \n",
       "2   chr11    2182062    2182385  AMPL7154784954  .   \n",
       "3   chr11    2182062    2182385  AMPL7154784954  .   \n",
       "4   chr11    2182130    2182458  AMPL7154784955  .   \n",
       "5   chr11    2182130    2182458  AMPL7154784955  .   \n",
       "6   chr11    2182429    2182753  AMPL7154897326  .   \n",
       "7   chr11    2182429    2182753  AMPL7154897326  .   \n",
       "8   chr12  121438967  121439264  AMPL7154784910  .   \n",
       "9   chr12  121438967  121439264  AMPL7154784910  .   \n",
       "10  chr12  121439250  121439535  AMPL7154784911  .   \n",
       "11  chr12  121439250  121439535  AMPL7154784911  .   \n",
       "12  chr12  121439524  121439785  AMPL7154784912  .   \n",
       "13  chr12  121439524  121439785  AMPL7154784912  .   \n",
       "14  chr12  121439774  121440004  AMPL7154783021  .   \n",
       "15  chr12  121439774  121440004  AMPL7154783021  .   \n",
       "16  chr12  121439987  121440308  AMPL7159709172  .   \n",
       "17  chr12  121439987  121440308  AMPL7159709172  .   \n",
       "\n",
       "                                                                       5  \\\n",
       "0   Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   \n",
       "1   Pool=1;SUBMITTED_REGION=AMPL7166063050,AMPL7154784954,AMPL7154784953   \n",
       "2   Pool=2;SUBMITTED_REGION=AMPL7154784954,AMPL7154784953,AMPL7154784955   \n",
       "3   Pool=2;SUBMITTED_REGION=AMPL7154784954,AMPL7154784953,AMPL7154784955   \n",
       "4   Pool=1;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326,AMPL7154784954   \n",
       "5   Pool=1;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326,AMPL7154784954   \n",
       "6                  Pool=2;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326   \n",
       "7                  Pool=2;SUBMITTED_REGION=AMPL7154784955,AMPL7154897326   \n",
       "8   Pool=1;SUBMITTED_REGION=AMPL7154784910,AMPL7154511317,AMPL7154784911   \n",
       "9   Pool=1;SUBMITTED_REGION=AMPL7154784910,AMPL7154511317,AMPL7154784911   \n",
       "10  Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154784910,AMPL7154784911   \n",
       "11  Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154784910,AMPL7154784911   \n",
       "12  Pool=1;SUBMITTED_REGION=AMPL7154784911,AMPL7154784912,AMPL7154783021   \n",
       "13  Pool=1;SUBMITTED_REGION=AMPL7154784911,AMPL7154784912,AMPL7154783021   \n",
       "14  Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154783021,AMPL7159709172   \n",
       "15  Pool=2;SUBMITTED_REGION=AMPL7154784912,AMPL7154783021,AMPL7159709172   \n",
       "16  Pool=1;SUBMITTED_REGION=AMPL7154783021,AMPL7159709205,AMPL7159709172   \n",
       "17  Pool=1;SUBMITTED_REGION=AMPL7154783021,AMPL7159709205,AMPL7159709172   \n",
       "\n",
       "        6                      7     8          9         10 11 12 13  \\\n",
       "0   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182439  .  -  .   \n",
       "1   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   \n",
       "2   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   \n",
       "3   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   \n",
       "4   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   \n",
       "5   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182218  .  -  .   \n",
       "6   chr11  ncbiRefSeq.2021-05-17  exon    2182015    2182439  .  -  .   \n",
       "7   chr11  ncbiRefSeq.2021-05-17  exon    2182398    2182439  .  -  .   \n",
       "8   chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   \n",
       "9   chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   \n",
       "10  chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   \n",
       "11  chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   \n",
       "12  chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   \n",
       "13  chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   \n",
       "14  chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   \n",
       "15  chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   \n",
       "16  chr12  ncbiRefSeq.2021-05-17  exon  121438289  121442292  .  -  .   \n",
       "17  chr12  ncbiRefSeq.2021-05-17  exon  121438868  121440315  .  +  .   \n",
       "\n",
       "          14  ampl_len  ref_len  ampl_start-ref_start  ref_end-ampl_end  \\\n",
       "0        INS       239      424                  -181               366   \n",
       "1   INS-IGF2       239      203                  -181               145   \n",
       "2        INS       323      203                    47              -167   \n",
       "3   INS-IGF2       323      203                    47              -167   \n",
       "4        INS       328      203                   115              -240   \n",
       "5   INS-IGF2       328      203                   115              -240   \n",
       "6        INS       324      424                   414              -314   \n",
       "7   INS-IGF2       324       41                    31              -314   \n",
       "8   C12orf43       297     4003                   678              3028   \n",
       "9      HNF1A       297     1447                    99              1051   \n",
       "10  C12orf43       285     4003                   961              2757   \n",
       "11     HNF1A       285     1447                   382               780   \n",
       "12  C12orf43       261     4003                  1235              2507   \n",
       "13     HNF1A       261     1447                   656               530   \n",
       "14  C12orf43       230     4003                  1485              2288   \n",
       "15     HNF1A       230     1447                   906               311   \n",
       "16  C12orf43       321     4003                  1698              1984   \n",
       "17     HNF1A       321     1447                  1119                 7   \n",
       "\n",
       "    overlap_len  \n",
       "0         -58.0  \n",
       "1         -58.0  \n",
       "2         156.0  \n",
       "3         156.0  \n",
       "4          88.0  \n",
       "5          88.0  \n",
       "6          10.0  \n",
       "7          10.0  \n",
       "8         297.0  \n",
       "9         297.0  \n",
       "10        285.0  \n",
       "11        285.0  \n",
       "12        261.0  \n",
       "13        261.0  \n",
       "14        230.0  \n",
       "15        230.0  \n",
       "16        321.0  \n",
       "17        321.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = drop_unique_length(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "INS-IGF2 is a read-through gene, which overlaps with gene INS at the 5' region, so it's not off-target region.   \n",
    "Apart from this, there are several amplicons, mapped on two different genes - *HNF1A* and *C12orf43*. Based on the descriptions of these genes (table in task1, file `RESULTS/genes_summary.csv`) I can assume, that *HNF1A* is a homologous off-target region.    \n",
    "I could be important to notice, that these genes have similar coordinates in different DNA strand. \n",
    "   \n",
    "**Target region coordinates:** 121438868, 121440315    \n",
    "**Homologous off-target coordinates:** 121438289, 121442292    \n",
    "   \n",
    "\n",
    "Save results:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.loc[df[14].isin(['HNF1A', 'C12orf43'])].reset_index().drop('index', axis=1)\n",
    "df.to_csv('RESULTS/task2_homologous_region.csv', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*******************************"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Two other ideas how to find homologous off-target region   \n",
    "3.1 [**Segmental Duplication Database**](https://eeelegacy.gs.washington.edu/humanparalogy/build37/build37.htm)[[3]](https://pubmed.ncbi.nlm.nih.gov/11381028/)     \n",
    "\n",
    "Download these database and try `bedtools intersect` again\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "sddb = pd.read_excel(\"build37.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>otherChrom</th>\n",
       "      <th>otherStart</th>\n",
       "      <th>otherEnd</th>\n",
       "      <th>otherSize</th>\n",
       "      <th>...</th>\n",
       "      <th>indelS</th>\n",
       "      <th>alignB</th>\n",
       "      <th>matchB</th>\n",
       "      <th>mismatchB</th>\n",
       "      <th>transitionsB</th>\n",
       "      <th>transversionsB</th>\n",
       "      <th>fracMatch</th>\n",
       "      <th>fracMatchIndel</th>\n",
       "      <th>jcK</th>\n",
       "      <th>k2K</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>85326</td>\n",
       "      <td>87112</td>\n",
       "      <td>chr1:398212</td>\n",
       "      <td>0</td>\n",
       "      <td>_</td>\n",
       "      <td>chr1</td>\n",
       "      <td>398212</td>\n",
       "      <td>400000</td>\n",
       "      <td>1788</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1786</td>\n",
       "      <td>1757</td>\n",
       "      <td>29</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>0.983763</td>\n",
       "      <td>0.982662</td>\n",
       "      <td>0.016416</td>\n",
       "      <td>0.016423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>398212</td>\n",
       "      <td>400000</td>\n",
       "      <td>chr1:85326</td>\n",
       "      <td>0</td>\n",
       "      <td>_</td>\n",
       "      <td>chr1</td>\n",
       "      <td>85326</td>\n",
       "      <td>87112</td>\n",
       "      <td>1786</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1786</td>\n",
       "      <td>1757</td>\n",
       "      <td>29</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>0.983763</td>\n",
       "      <td>0.982662</td>\n",
       "      <td>0.016416</td>\n",
       "      <td>0.016423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>88000</td>\n",
       "      <td>121417</td>\n",
       "      <td>chr1:235525</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>235525</td>\n",
       "      <td>267707</td>\n",
       "      <td>32182</td>\n",
       "      <td>...</td>\n",
       "      <td>1299</td>\n",
       "      <td>32150</td>\n",
       "      <td>31941</td>\n",
       "      <td>209</td>\n",
       "      <td>133</td>\n",
       "      <td>76</td>\n",
       "      <td>0.993499</td>\n",
       "      <td>0.992727</td>\n",
       "      <td>0.006529</td>\n",
       "      <td>0.006532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>235525</td>\n",
       "      <td>267707</td>\n",
       "      <td>chr1:88000</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>88000</td>\n",
       "      <td>121417</td>\n",
       "      <td>33417</td>\n",
       "      <td>...</td>\n",
       "      <td>1299</td>\n",
       "      <td>32150</td>\n",
       "      <td>31941</td>\n",
       "      <td>209</td>\n",
       "      <td>133</td>\n",
       "      <td>76</td>\n",
       "      <td>0.993499</td>\n",
       "      <td>0.992727</td>\n",
       "      <td>0.006529</td>\n",
       "      <td>0.006532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91256</td>\n",
       "      <td>92392</td>\n",
       "      <td>chr1:521369</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>521369</td>\n",
       "      <td>522487</td>\n",
       "      <td>1118</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>1117</td>\n",
       "      <td>1092</td>\n",
       "      <td>25</td>\n",
       "      <td>18</td>\n",
       "      <td>7</td>\n",
       "      <td>0.977619</td>\n",
       "      <td>0.974130</td>\n",
       "      <td>0.022722</td>\n",
       "      <td>0.022781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51594</th>\n",
       "      <td>chrY</td>\n",
       "      <td>58880431</td>\n",
       "      <td>58891092</td>\n",
       "      <td>chrY:58869728</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>58869728</td>\n",
       "      <td>58880430</td>\n",
       "      <td>10702</td>\n",
       "      <td>...</td>\n",
       "      <td>99</td>\n",
       "      <td>10632</td>\n",
       "      <td>10566</td>\n",
       "      <td>66</td>\n",
       "      <td>25</td>\n",
       "      <td>41</td>\n",
       "      <td>0.993792</td>\n",
       "      <td>0.992485</td>\n",
       "      <td>0.006234</td>\n",
       "      <td>0.006234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51595</th>\n",
       "      <td>chrY</td>\n",
       "      <td>58881608</td>\n",
       "      <td>58885137</td>\n",
       "      <td>chrY:58885137</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>58885137</td>\n",
       "      <td>58888721</td>\n",
       "      <td>3584</td>\n",
       "      <td>...</td>\n",
       "      <td>55</td>\n",
       "      <td>3529</td>\n",
       "      <td>3509</td>\n",
       "      <td>20</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>0.994333</td>\n",
       "      <td>0.993488</td>\n",
       "      <td>0.005689</td>\n",
       "      <td>0.005689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51596</th>\n",
       "      <td>chrY</td>\n",
       "      <td>58885137</td>\n",
       "      <td>58888721</td>\n",
       "      <td>chrY:58881608</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>58881608</td>\n",
       "      <td>58885137</td>\n",
       "      <td>3529</td>\n",
       "      <td>...</td>\n",
       "      <td>55</td>\n",
       "      <td>3529</td>\n",
       "      <td>3509</td>\n",
       "      <td>20</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>0.994333</td>\n",
       "      <td>0.993488</td>\n",
       "      <td>0.005689</td>\n",
       "      <td>0.005689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51597</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59012496</td>\n",
       "      <td>59013612</td>\n",
       "      <td>chrY:59024857</td>\n",
       "      <td>0</td>\n",
       "      <td>_</td>\n",
       "      <td>chrY</td>\n",
       "      <td>59024857</td>\n",
       "      <td>59025974</td>\n",
       "      <td>1117</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1115</td>\n",
       "      <td>1041</td>\n",
       "      <td>74</td>\n",
       "      <td>42</td>\n",
       "      <td>32</td>\n",
       "      <td>0.933632</td>\n",
       "      <td>0.931961</td>\n",
       "      <td>0.069490</td>\n",
       "      <td>0.069706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51598</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59024857</td>\n",
       "      <td>59025974</td>\n",
       "      <td>chrY:59012496</td>\n",
       "      <td>0</td>\n",
       "      <td>_</td>\n",
       "      <td>chrY</td>\n",
       "      <td>59012496</td>\n",
       "      <td>59013612</td>\n",
       "      <td>1116</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1115</td>\n",
       "      <td>1041</td>\n",
       "      <td>74</td>\n",
       "      <td>42</td>\n",
       "      <td>32</td>\n",
       "      <td>0.933632</td>\n",
       "      <td>0.931961</td>\n",
       "      <td>0.069490</td>\n",
       "      <td>0.069706</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51599 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  chromStart  chromEnd           name  score strand otherChrom  \\\n",
       "0      chr1       85326     87112    chr1:398212      0      _       chr1   \n",
       "1      chr1      398212    400000     chr1:85326      0      _       chr1   \n",
       "2      chr1       88000    121417    chr1:235525      0      +       chr1   \n",
       "3      chr1      235525    267707     chr1:88000      0      +       chr1   \n",
       "4      chr1       91256     92392    chr1:521369      0      +       chr1   \n",
       "...     ...         ...       ...            ...    ...    ...        ...   \n",
       "51594  chrY    58880431  58891092  chrY:58869728      0      +       chrY   \n",
       "51595  chrY    58881608  58885137  chrY:58885137      0      +       chrY   \n",
       "51596  chrY    58885137  58888721  chrY:58881608      0      +       chrY   \n",
       "51597  chrY    59012496  59013612  chrY:59024857      0      _       chrY   \n",
       "51598  chrY    59024857  59025974  chrY:59012496      0      _       chrY   \n",
       "\n",
       "       otherStart  otherEnd  otherSize  ...  indelS  alignB  matchB  \\\n",
       "0          398212    400000       1788  ...       2    1786    1757   \n",
       "1           85326     87112       1786  ...       2    1786    1757   \n",
       "2          235525    267707      32182  ...    1299   32150   31941   \n",
       "3           88000    121417      33417  ...    1299   32150   31941   \n",
       "4          521369    522487       1118  ...      20    1117    1092   \n",
       "...           ...       ...        ...  ...     ...     ...     ...   \n",
       "51594    58869728  58880430      10702  ...      99   10632   10566   \n",
       "51595    58885137  58888721       3584  ...      55    3529    3509   \n",
       "51596    58881608  58885137       3529  ...      55    3529    3509   \n",
       "51597    59024857  59025974       1117  ...       3    1115    1041   \n",
       "51598    59012496  59013612       1116  ...       3    1115    1041   \n",
       "\n",
       "       mismatchB  transitionsB  transversionsB fracMatch  fracMatchIndel  \\\n",
       "0             29            15              14  0.983763        0.982662   \n",
       "1             29            15              14  0.983763        0.982662   \n",
       "2            209           133              76  0.993499        0.992727   \n",
       "3            209           133              76  0.993499        0.992727   \n",
       "4             25            18               7  0.977619        0.974130   \n",
       "...          ...           ...             ...       ...             ...   \n",
       "51594         66            25              41  0.993792        0.992485   \n",
       "51595         20             9              11  0.994333        0.993488   \n",
       "51596         20             9              11  0.994333        0.993488   \n",
       "51597         74            42              32  0.933632        0.931961   \n",
       "51598         74            42              32  0.933632        0.931961   \n",
       "\n",
       "            jcK       k2K  \n",
       "0      0.016416  0.016423  \n",
       "1      0.016416  0.016423  \n",
       "2      0.006529  0.006532  \n",
       "3      0.006529  0.006532  \n",
       "4      0.022722  0.022781  \n",
       "...         ...       ...  \n",
       "51594  0.006234  0.006234  \n",
       "51595  0.005689  0.005689  \n",
       "51596  0.005689  0.005689  \n",
       "51597  0.069490  0.069706  \n",
       "51598  0.069490  0.069706  \n",
       "\n",
       "[51599 rows x 29 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sddb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extract 4-column BED-file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "sddb.rename(columns={'chrom':'#chrom'})\n",
    "sddb_short = sddb.iloc[:, [0, 1, 2, 3]]\n",
    "sddb_short.to_csv('sddb.bed', index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run `bedtools intersect` with SDDB BED-file and output BED-file from tak1 (to have genes names):   \n",
    "```bash\n",
    "bedtools intersect -a ./RESULTS/task1_output.bed -b sddb.bed -wa -wb > sddb_intersect.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135937315</td>\n",
       "      <td>135937599</td>\n",
       "      <td>AMPL7154848171</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154848171</td>\n",
       "      <td>CEL</td>\n",
       "      <td>1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135939554</td>\n",
       "      <td>135939855</td>\n",
       "      <td>AMPL7154785491</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154785491,AMPL7162958722</td>\n",
       "      <td>CEL</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135939844</td>\n",
       "      <td>135940096</td>\n",
       "      <td>AMPL7162958722</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154783210,AMPL7154785491,AMPL7162958722</td>\n",
       "      <td>CEL</td>\n",
       "      <td>2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135939844</td>\n",
       "      <td>135940096</td>\n",
       "      <td>AMPL7162958722</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154783210,AMPL7154785491,AMPL7162958722</td>\n",
       "      <td>CEL</td>\n",
       "      <td>3</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135940061</td>\n",
       "      <td>135940234</td>\n",
       "      <td>AMPL7154783210</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154785486,AMPL7154783210,AMPL7162958722</td>\n",
       "      <td>CEL</td>\n",
       "      <td>3</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135940223</td>\n",
       "      <td>135940552</td>\n",
       "      <td>AMPL7154785486</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154785486,AMPL7154783210,AMPL7154848184</td>\n",
       "      <td>CEL</td>\n",
       "      <td>4</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135940524</td>\n",
       "      <td>135940722</td>\n",
       "      <td>AMPL7154848184</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154848184,AMPL7154785486</td>\n",
       "      <td>CEL</td>\n",
       "      <td>4</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135941786</td>\n",
       "      <td>135942083</td>\n",
       "      <td>AMPL7154848187</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154848187</td>\n",
       "      <td>CEL</td>\n",
       "      <td>5</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135942100</td>\n",
       "      <td>135942411</td>\n",
       "      <td>AMPL7154848190</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154848193,AMPL7154848190</td>\n",
       "      <td>CEL</td>\n",
       "      <td>6</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135942385</td>\n",
       "      <td>135942705</td>\n",
       "      <td>AMPL7154848193</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154848190,AMPL7154848193</td>\n",
       "      <td>CEL</td>\n",
       "      <td>7</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135943934</td>\n",
       "      <td>135944214</td>\n",
       "      <td>AMPL7154785490</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166072477,AMPL7154785490</td>\n",
       "      <td>CEL</td>\n",
       "      <td>8</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135944183</td>\n",
       "      <td>135944500</td>\n",
       "      <td>AMPL7166072477</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166072477,AMPL7154785490,AMPL7166072478</td>\n",
       "      <td>CEL</td>\n",
       "      <td>8</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135944183</td>\n",
       "      <td>135944500</td>\n",
       "      <td>AMPL7166072477</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166072477,AMPL7154785490,AMPL7166072478</td>\n",
       "      <td>CEL</td>\n",
       "      <td>9</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135944485</td>\n",
       "      <td>135944807</td>\n",
       "      <td>AMPL7166072478</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7166072477,AMPL7166072478</td>\n",
       "      <td>CEL</td>\n",
       "      <td>9</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135945694</td>\n",
       "      <td>135946019</td>\n",
       "      <td>AMPL7154785492</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154785492,AMPL7162954932</td>\n",
       "      <td>CEL</td>\n",
       "      <td>10</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135946008</td>\n",
       "      <td>135946189</td>\n",
       "      <td>AMPL7162954932</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7154785581,AMPL7162954932,AMPL7154785492</td>\n",
       "      <td>CEL</td>\n",
       "      <td>10</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135946180</td>\n",
       "      <td>135946510</td>\n",
       "      <td>AMPL7154785581</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7154785581,AMPL7162954932</td>\n",
       "      <td>CEL</td>\n",
       "      <td>11</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr9</td>\n",
       "      <td>135947121</td>\n",
       "      <td>135947302</td>\n",
       "      <td>AMPL7163451433</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=2;SUBMITTED_REGION=AMPL7163451433</td>\n",
       "      <td>CEL</td>\n",
       "      <td>11</td>\n",
       "      <td>chr9</td>\n",
       "      <td>135935536</td>\n",
       "      <td>135947286</td>\n",
       "      <td>chr9:135956012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr10</td>\n",
       "      <td>88811251</td>\n",
       "      <td>88811541</td>\n",
       "      <td>AMPL7166063029</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063029</td>\n",
       "      <td>GLUD1</td>\n",
       "      <td>16</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88810245</td>\n",
       "      <td>88811564</td>\n",
       "      <td>chr10:31916223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr10</td>\n",
       "      <td>88811251</td>\n",
       "      <td>88811541</td>\n",
       "      <td>AMPL7166063029</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063029</td>\n",
       "      <td>GLUD1</td>\n",
       "      <td>16</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88810248</td>\n",
       "      <td>88811628</td>\n",
       "      <td>chrX:120183094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr10</td>\n",
       "      <td>88811251</td>\n",
       "      <td>88811541</td>\n",
       "      <td>AMPL7166063029</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063029</td>\n",
       "      <td>GLUD1</td>\n",
       "      <td>15</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88810245</td>\n",
       "      <td>88811564</td>\n",
       "      <td>chr10:31916223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr10</td>\n",
       "      <td>88811251</td>\n",
       "      <td>88811541</td>\n",
       "      <td>AMPL7166063029</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063029</td>\n",
       "      <td>GLUD1</td>\n",
       "      <td>15</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88810248</td>\n",
       "      <td>88811628</td>\n",
       "      <td>chrX:120183094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr10</td>\n",
       "      <td>88811251</td>\n",
       "      <td>88811541</td>\n",
       "      <td>AMPL7166063029</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063029</td>\n",
       "      <td>GLUD1</td>\n",
       "      <td>13</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88810245</td>\n",
       "      <td>88811564</td>\n",
       "      <td>chr10:31916223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr10</td>\n",
       "      <td>88811251</td>\n",
       "      <td>88811541</td>\n",
       "      <td>AMPL7166063029</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063029</td>\n",
       "      <td>GLUD1</td>\n",
       "      <td>13</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88810248</td>\n",
       "      <td>88811628</td>\n",
       "      <td>chrX:120183094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr10</td>\n",
       "      <td>88811251</td>\n",
       "      <td>88811541</td>\n",
       "      <td>AMPL7166063029</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063029</td>\n",
       "      <td>GLUD1</td>\n",
       "      <td>14</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88810245</td>\n",
       "      <td>88811564</td>\n",
       "      <td>chr10:31916223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr10</td>\n",
       "      <td>88811251</td>\n",
       "      <td>88811541</td>\n",
       "      <td>AMPL7166063029</td>\n",
       "      <td>.</td>\n",
       "      <td>Pool=1;SUBMITTED_REGION=AMPL7166063029</td>\n",
       "      <td>GLUD1</td>\n",
       "      <td>14</td>\n",
       "      <td>chr10</td>\n",
       "      <td>88810248</td>\n",
       "      <td>88811628</td>\n",
       "      <td>chrX:120183094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0          1          2               3  4   \\\n",
       "0    chr9  135937315  135937599  AMPL7154848171  .   \n",
       "1    chr9  135939554  135939855  AMPL7154785491  .   \n",
       "2    chr9  135939844  135940096  AMPL7162958722  .   \n",
       "3    chr9  135939844  135940096  AMPL7162958722  .   \n",
       "4    chr9  135940061  135940234  AMPL7154783210  .   \n",
       "5    chr9  135940223  135940552  AMPL7154785486  .   \n",
       "6    chr9  135940524  135940722  AMPL7154848184  .   \n",
       "7    chr9  135941786  135942083  AMPL7154848187  .   \n",
       "8    chr9  135942100  135942411  AMPL7154848190  .   \n",
       "9    chr9  135942385  135942705  AMPL7154848193  .   \n",
       "10   chr9  135943934  135944214  AMPL7154785490  .   \n",
       "11   chr9  135944183  135944500  AMPL7166072477  .   \n",
       "12   chr9  135944183  135944500  AMPL7166072477  .   \n",
       "13   chr9  135944485  135944807  AMPL7166072478  .   \n",
       "14   chr9  135945694  135946019  AMPL7154785492  .   \n",
       "15   chr9  135946008  135946189  AMPL7162954932  .   \n",
       "16   chr9  135946180  135946510  AMPL7154785581  .   \n",
       "17   chr9  135947121  135947302  AMPL7163451433  .   \n",
       "18  chr10   88811251   88811541  AMPL7166063029  .   \n",
       "19  chr10   88811251   88811541  AMPL7166063029  .   \n",
       "20  chr10   88811251   88811541  AMPL7166063029  .   \n",
       "21  chr10   88811251   88811541  AMPL7166063029  .   \n",
       "22  chr10   88811251   88811541  AMPL7166063029  .   \n",
       "23  chr10   88811251   88811541  AMPL7166063029  .   \n",
       "24  chr10   88811251   88811541  AMPL7166063029  .   \n",
       "25  chr10   88811251   88811541  AMPL7166063029  .   \n",
       "\n",
       "                                                                      5   \\\n",
       "0                                 Pool=1;SUBMITTED_REGION=AMPL7154848171   \n",
       "1                  Pool=2;SUBMITTED_REGION=AMPL7154785491,AMPL7162958722   \n",
       "2   Pool=1;SUBMITTED_REGION=AMPL7154783210,AMPL7154785491,AMPL7162958722   \n",
       "3   Pool=1;SUBMITTED_REGION=AMPL7154783210,AMPL7154785491,AMPL7162958722   \n",
       "4   Pool=2;SUBMITTED_REGION=AMPL7154785486,AMPL7154783210,AMPL7162958722   \n",
       "5   Pool=1;SUBMITTED_REGION=AMPL7154785486,AMPL7154783210,AMPL7154848184   \n",
       "6                  Pool=2;SUBMITTED_REGION=AMPL7154848184,AMPL7154785486   \n",
       "7                                 Pool=1;SUBMITTED_REGION=AMPL7154848187   \n",
       "8                  Pool=2;SUBMITTED_REGION=AMPL7154848193,AMPL7154848190   \n",
       "9                  Pool=1;SUBMITTED_REGION=AMPL7154848190,AMPL7154848193   \n",
       "10                 Pool=2;SUBMITTED_REGION=AMPL7166072477,AMPL7154785490   \n",
       "11  Pool=1;SUBMITTED_REGION=AMPL7166072477,AMPL7154785490,AMPL7166072478   \n",
       "12  Pool=1;SUBMITTED_REGION=AMPL7166072477,AMPL7154785490,AMPL7166072478   \n",
       "13                 Pool=2;SUBMITTED_REGION=AMPL7166072477,AMPL7166072478   \n",
       "14                 Pool=1;SUBMITTED_REGION=AMPL7154785492,AMPL7162954932   \n",
       "15  Pool=2;SUBMITTED_REGION=AMPL7154785581,AMPL7162954932,AMPL7154785492   \n",
       "16                 Pool=1;SUBMITTED_REGION=AMPL7154785581,AMPL7162954932   \n",
       "17                                Pool=2;SUBMITTED_REGION=AMPL7163451433   \n",
       "18                                Pool=1;SUBMITTED_REGION=AMPL7166063029   \n",
       "19                                Pool=1;SUBMITTED_REGION=AMPL7166063029   \n",
       "20                                Pool=1;SUBMITTED_REGION=AMPL7166063029   \n",
       "21                                Pool=1;SUBMITTED_REGION=AMPL7166063029   \n",
       "22                                Pool=1;SUBMITTED_REGION=AMPL7166063029   \n",
       "23                                Pool=1;SUBMITTED_REGION=AMPL7166063029   \n",
       "24                                Pool=1;SUBMITTED_REGION=AMPL7166063029   \n",
       "25                                Pool=1;SUBMITTED_REGION=AMPL7166063029   \n",
       "\n",
       "       6   7      8          9          10              11  \n",
       "0     CEL   1   chr9  135935536  135947286  chr9:135956012  \n",
       "1     CEL   2   chr9  135935536  135947286  chr9:135956012  \n",
       "2     CEL   2   chr9  135935536  135947286  chr9:135956012  \n",
       "3     CEL   3   chr9  135935536  135947286  chr9:135956012  \n",
       "4     CEL   3   chr9  135935536  135947286  chr9:135956012  \n",
       "5     CEL   4   chr9  135935536  135947286  chr9:135956012  \n",
       "6     CEL   4   chr9  135935536  135947286  chr9:135956012  \n",
       "7     CEL   5   chr9  135935536  135947286  chr9:135956012  \n",
       "8     CEL   6   chr9  135935536  135947286  chr9:135956012  \n",
       "9     CEL   7   chr9  135935536  135947286  chr9:135956012  \n",
       "10    CEL   8   chr9  135935536  135947286  chr9:135956012  \n",
       "11    CEL   8   chr9  135935536  135947286  chr9:135956012  \n",
       "12    CEL   9   chr9  135935536  135947286  chr9:135956012  \n",
       "13    CEL   9   chr9  135935536  135947286  chr9:135956012  \n",
       "14    CEL  10   chr9  135935536  135947286  chr9:135956012  \n",
       "15    CEL  10   chr9  135935536  135947286  chr9:135956012  \n",
       "16    CEL  11   chr9  135935536  135947286  chr9:135956012  \n",
       "17    CEL  11   chr9  135935536  135947286  chr9:135956012  \n",
       "18  GLUD1  16  chr10   88810245   88811564  chr10:31916223  \n",
       "19  GLUD1  16  chr10   88810248   88811628  chrX:120183094  \n",
       "20  GLUD1  15  chr10   88810245   88811564  chr10:31916223  \n",
       "21  GLUD1  15  chr10   88810248   88811628  chrX:120183094  \n",
       "22  GLUD1  13  chr10   88810245   88811564  chr10:31916223  \n",
       "23  GLUD1  13  chr10   88810248   88811628  chrX:120183094  \n",
       "24  GLUD1  14  chr10   88810245   88811564  chr10:31916223  \n",
       "25  GLUD1  14  chr10   88810248   88811628  chrX:120183094  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('sddb_intersect.txt', sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see, that regions in chr10 have duplicates in chrX, for example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>otherChrom</th>\n",
       "      <th>otherStart</th>\n",
       "      <th>otherEnd</th>\n",
       "      <th>otherSize</th>\n",
       "      <th>...</th>\n",
       "      <th>indelS</th>\n",
       "      <th>alignB</th>\n",
       "      <th>matchB</th>\n",
       "      <th>mismatchB</th>\n",
       "      <th>transitionsB</th>\n",
       "      <th>transversionsB</th>\n",
       "      <th>fracMatch</th>\n",
       "      <th>fracMatchIndel</th>\n",
       "      <th>jcK</th>\n",
       "      <th>k2K</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8705</th>\n",
       "      <td>chr10</td>\n",
       "      <td>88810248</td>\n",
       "      <td>88811628</td>\n",
       "      <td>chrX:120183094</td>\n",
       "      <td>0</td>\n",
       "      <td>_</td>\n",
       "      <td>chrX</td>\n",
       "      <td>120183094</td>\n",
       "      <td>120184459</td>\n",
       "      <td>1365</td>\n",
       "      <td>...</td>\n",
       "      <td>31</td>\n",
       "      <td>1357</td>\n",
       "      <td>1308</td>\n",
       "      <td>49</td>\n",
       "      <td>36</td>\n",
       "      <td>13</td>\n",
       "      <td>0.963891</td>\n",
       "      <td>0.95684</td>\n",
       "      <td>0.037007</td>\n",
       "      <td>0.037179</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  chromStart  chromEnd            name  score strand otherChrom  \\\n",
       "8705  chr10    88810248  88811628  chrX:120183094      0      _       chrX   \n",
       "\n",
       "      otherStart   otherEnd  otherSize  ...  indelS  alignB  matchB  \\\n",
       "8705   120183094  120184459       1365  ...      31    1357    1308   \n",
       "\n",
       "      mismatchB  transitionsB  transversionsB fracMatch  fracMatchIndel  \\\n",
       "8705         49            36              13  0.963891         0.95684   \n",
       "\n",
       "           jcK       k2K  \n",
       "8705  0.037007  0.037179  \n",
       "\n",
       "[1 rows x 29 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sddb.loc[(sddb['chrom'] == 'chr10') & (sddb['chromStart'] == 88810248)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But there are indels. Perhaps, with a more detailed approach, this database could be useful in searching homologous off-target regions.   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3.2 **Special tools for searching homologous regions**   \n",
    "There are some tools and pipelines used for searching homologous region, such as [i-ADHoRe](https://academic.oup.com/nar/article/40/2/e11/2409742), [FASTA](https://www.ebi.ac.uk/Tools/sss/fasta/nucleotide.html) or [this algorithm](https://www.nature.com/articles/gim201658)   \n",
    "   \n",
    "All of them work with fasta files, we can get it easily from BED-file.   \n",
    "First of all, we need to get fasta file from input BED-file.  \n",
    "    \n",
    "Download reference genome hg19:   \n",
    "```bash\n",
    "wget ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/*\n",
    "for file in ./*.gz; do gunzip $file; done\n",
    "cat *.fa > hg19.genome.fa\n",
    "```\n",
    "   \n",
    "Getting fasta from BED:   \n",
    "```bash\n",
    "bedtools getfasta -fi hg19.genome.fa -bed IAD143293_241_Designed.bed > input.fasta\n",
    "```\n",
    "   \n",
    "I am not testing these tools, but I assume, if we understand their work well, we can adapt them to our needs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*******************"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
